---
title: "subpopulatoin analysis"
output: html_notebook
---

# Summary
This notebook performs simple exploratory data analysis for single-cell EMT data. 
Some of the code was adapted from various sources.

# 1. Download all libraries and library dependencies
First, we'll download several libraries and library dependencies for single-cell
data processing and data visualization.

Note that the following codeblock takes awhile to run. Only do it if you do not
have the libraries pre-installed in your R environment.
```{r, include=TRUE}
#pkgLst = c("Seurat", "dplyr", "viridis", 
#           "matrixStats", "UpSetR", 
#           "mgcv", "tidyverse", "reshape2", "ggplot2")
#install.packages(pkgLst)

library(mgcv)
library(splines)
library(Seurat)
library(dplyr)
library(tidyr)
library(viridis)
library(matrixStats)
library(UpSetR)
library(reshape2)
library(ggplot2)
library(Rmagic)
library(readxl)
```

# 2. Load A549 TGF-B dataset
First, let's load up the data from A549.
```{r, include=TRUE}
rm(list=ls())
#datapath = "C:/Users/scott/Data/scRNASeq/GSE147405/A549/A549_TGFB1.rds"
datapath = "D:/Data/scRNASeq/GSE147405/A549/A549_TGFB1.rds"
a549_tgfb = readRDS(datapath)
unique(a549_tgfb@meta.data$Time)
```
## Filter genes
Let's filter out some low quality genes.
```{r}
genes_filter = rownames(a549_tgfb[["RNA"]]@data)[rowSums(as.matrix(a549_tgfb[["RNA"]]@data)) > 0]
a549_tgfb = subset(a549_tgfb, features=genes_filter)
```

## Regress out cell cycle
We can regress some biological variance out, including cell cycle genes.
```{r}
genes_to_regress = "D:/Data/scRNASeq/GSE147405/regev_lab_cell_cycle_genes.txt"
cc.genes = readLines(con=genes_to_regress)
s.genes  = cc.genes[1:43]
g2m.genes = cc.genes[44:97]
a549_tgfb = CellCycleScoring(a549_tgfb, 
                             s.features=s.genes,
                             g2m.features=g2m.genes, 
                             set.ident=FALSE)
```

## Subset time points of interest
Finally, let's subset out some of the cells.
```{r}
a549_tgfb = subset(a549_tgfb, subset = Time==c("0d", "8h", "3d", "1d", "7d"))
orig_id = colnames(a549_tgfb@assays$RNA@counts)
all_ids = cbind(orig_id, a549_tgfb@meta.data$Time)

savepath = "D:/Analysis/EMT/a549_colmap.csv"
write.csv(all_ids, savepath)
```

## Preprocess data for MAGIC
Now, we'll run MAGIC on the dataset.
```{r}
data = data.frame(a549_tgfb@assays$RNA@counts)
data = t(data)
```

Ensure our initial filtering worked.
```{R}
# keep genes expressed in at least 10 cells
keep_cols = colSums(data > 0) > 10
data = data[,keep_cols]
```

Plot the data distribution.
```{r}
# look at the distribution of library sizes
ggplot() +
  geom_histogram(aes(x=rowSums(data)), bins=50) +
  geom_vline(xintercept = 1000, color='red')
```

Normalize by library size and take square root.
```{r}
data = library.size.normalize(data)
data = sqrt(data)
```

After this, all we have to do is run the dataset through MAGIC in Python.
```{r}
write.table(data, file="D:/Analysis/EMT/a549_tgfb_preprocessed_for_magic.csv", 
            row.names=TRUE, col.names=TRUE)
```

# 3. Data preprocessing
```{r, results='hide'}
find_subpopulations = function(gene_obj){
  
  # Identify variable features
  gene_obj = FindVariableFeatures(object=gene_obj, 
                              mean.function=ExpMean,
                              dispersion.function=LogVMR,
                              selection.method='vst',
                              x.low.cutoff=-Inf,
                              x.high.cutoff=Inf,
                              y.cutoff=Inf,
                              nfeatures = 2000)
  
  gene_obj <- SCTransform(gene_obj,
                         do.correct.umi=T,
                         vars.to.regress=c("S.Score", "G2M.Score",
                                           "percent.mito"),
                         #batch_var="Mix",
                         return_gene_attr=T,
                         variable.features.n=2000)

  
  # Run PCA
  gene_obj = RunPCA(object=gene_obj,
                verbose=FALSE)
  
  # Find shortest nearest neighbors and clusters
  gene_obj = FindNeighbors(gene_obj, 
                       dims=1:30,
                       reduction='pca')
  gene_obj = FindClusters(gene_obj, 
                      resolution=0.5)
  
  # Perform T-SNE
  gene_obj = RunTSNE(gene_obj,
                     dims=1:30)
  
  # Perform UMAP
  gene_obj = RunUMAP(gene_obj, 
                 dims=1:30)
  
  return(gene_obj)
}
```

```{r}
a549_tgfb = find_subpopulations(a549_tgfb)
cluster_pred = a549_tgfb@active.ident
```

```{r}
a549_tgfb@active.ident = as.factor(a549_tgfb@meta.data$Time)
DimPlot(a549_tgfb, 
        reduction="umap",
        dims=c(1, 2),
        #cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="", 
             x="UMAP 1", 
             y="UMAP 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.15),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))

a549_tgfb@active.ident = cluster_pred
DimPlot(a549_tgfb, 
        reduction="umap",
        dims=c(1, 2),
        #cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="", 
             x="UMAP 1", 
             y="UMAP 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.15),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))
```

# 4. Load metabolic flux data
We'll read in the metabolic flux data computed from single-cell Flux Balance Analysis
```{r, results='hide'}
path_to_neutrophil_flux = 'D:/Analysis/EMT/recon1_cobra_results.xlsx'
sheet_name = 'Flux'
flux_data = read_excel(path_to_neutrophil_flux, sheet_name)
rxn_ids = flux_data[, 1:3]
flux_data[, 1:3] = NULL
row.names(flux_data) = rxn_ids$ID
flux_data = t(flux_data)

flux_data = data.frame(flux_data)
#flux_data[, 2:ncol(flux_data)] = abs(flux_data[, 2:ncol(flux_data)])
```

```{r}
# Read in original data
path_to_old_ids = "D:/Analysis/EMT/a549_colmap.csv"
ids =read_csv(path_to_old_ids)
```

First, we'll grab the day number,
```{r, results='hide'}
# Grab day
ids$day = ids$X3
ids$day = gsub("[^0-9.-]", "", ids$day)
```

Then add a count to the end of that particular time point. Also, reformat it so that it looks like the old mapping.
```{r, results='hide'}
# Iteratively add 1 for all duplicates
ids$day = as.numeric(ids$day)
ids$day = ids$day * 10
ids = data.frame(ids)
ids$new_day = ave(ids$day, ids$X3, FUN = function(x) x[1] + seq_along(x) - 1)
ids$new_day = ids$new_day - ids$day
ids$day = ids$day / 10
```

Create the final identifier that perfectly imitates the old identifier, and manually set the first occurrences.
```{r, results='hide'}
# Create the new ID
ids$final_day = paste(ids$X3, ids$new_day, sep='.')
ids$final_day = as.character(ids$final_day)

# Manually set the first call for a given day
ids$final_day[1] = '3d'
ids$final_day[8] = '1d'
ids$final_day[10] = '7d'
ids$final_day[143] = '8h'
ids$final_day[96] = '0d'
```

We have to now merge the ids and the metabolic flux data to get a cohesive mapping of cell identifiers to flux.
```{r, results='hide'}
mapped_flux = merge(ids, 
                    flux_data, 
                    by.x='final_day', 
                    by.y='row.names')
```

## A. Format FBA Data
First, we need to format the data, so that it matches the DGE Matrix, where rows are reactions and columns are cell IDs.
```{r, results='hide'}
fba = mapped_flux

# Remove unnecessary columns
fba[, c("final_col", "day", "new_day", "final_day")] = NULL
fba[, c("X1", "X3")] = NULL
# Map to original IDs
rownames(fba) = fba$orig_id
fba$orig_id = NULL

# Reformat
fba = t(fba)
fba = as.sparse(fba)
```

Now we need to make sure the expression matrix has the same cells as the fba matrix, and also ensure the `active.ident` and the `Time` datapoints we'll map to.
```{r, results='hide'}
# Get names and indices
intersect_cells = intersect(colnames(a549_tgfb@assays$RNA@counts), colnames(fba))
cell_idx = which(colnames(a549_tgfb@assays$RNA@counts) %in% intersect_cells)

# Reformat data
a549_tgfb = subset(a549_tgfb, cells=intersect_cells)
```

## B. Create FBA Assay in Seurat Object
This incorporates FBA into the Seurat Object.
```{r, results='hide'}
fba = fba[, intersect_cells]
a549_tgfb[["FBA"]] = CreateAssayObject(counts=fba)
```

## C. Perform data preprocessing for metabolic fluxes
We'll now perform some data preprocessing steps for the metabolic fluxes. Let's define the function with the following preprocessing steps:
  1. Log normalization
  2. Data scaling

```{r, results='hide'}
preprocess_flux = function(gene_obj, assay){
  # Log normalize data
  #gene_obj = NormalizeData(gene_obj, 
  #                         assay=assay,
  #                         normalization.method='LogNormalize', 
  #                         scale.factor=10000)
  
  # Scale data
  #gene_obj = ScaleData(gene_obj, 
  #                     assay=assay)
  
  gene_obj = SCTransform(gene_obj)
  return(gene_obj)
}
```

Now let's perform the data preprocessing:
```{r, results='hide'}
a549_tgfb = preprocess_flux(a549_tgfb, assay="FBA")
```

## D. Get Metabolic Markers
We'll get metabolic markers based on subpopulations.
```{r, results='hide'}
a549_tgfb@active.ident = as.factor(a549_tgfb@meta.data$Time)
subpop_reactions = FindAllMarkers(a549_tgfb,
                                  assay="RNA")
#write.table(subpop_reactions, "clipboard", sep="\t")
```

Let's intialize some variables we need to declare for the visualization
```{r, results='hide'}
reactions_to_visualize = c("HMR_4418", "HMR_4574", "INSK",
                           "HMR_8486", "HMR_4012", "HMR_4085")
reaction_names = c("Hypoxanthine Phosphoribosyltransferase", 
                   "Inosine Phosphorylase", 
                   "Inosine Kinase",
                   "Adenosine Phosphorylase",
                   "Adenosine Kinase",
                   "Adenine Phosphoribosyltransferase")
colors = c("#CC3311", "#CC3311", "#CC3311",
           "#004488", "#004488", "#004488")
```

Create each T-SNE plot and store them in a list. 
```{r, results='hide'}
p = list()
for(i in 1:length(reactions_to_visualize)){
  p[[i]] = ggplot(merged_tsne, aes(x=tSNE_1, y=tSNE_2, color=merged_tsne[, reactions_to_visualize[i]])) + 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="t-SNE 1", 
             y="t-SNE 2",
             title=reaction_names[i], 
             color="Flux") + 
        scale_color_gradient(low="gray", high=colors[i]) +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())
}
```

Create and visualize the final T-SNE flux plot object.
```{r, fig.width=12, fig.height=6, results='hide'}
pC = scater::multiplot(p[[1]], p[[6]], p[[2]], p[[4]], p[[3]], p[[5]], cols=3)
pC
```

Finally, let's assemble the final figure:
```{r, results='hide'}
pX = ggarrange(pA, pB, 
               ncol=1, nrow=2, 
               labels=c("A", "B"))
pY = ggarrange(pC, 
               ncol=1, 
               labels="C",
               common.legend=FALSE)
ggarrange(pX, pY, 
          nrow=1, ncol=2, 
          widths=c(6, 24), heights=c(12, 12),
          common.legend=FALSE)
```

## E. Plot heatmap of differentially active metabolic reactions from T-SNE Neutrophil Subpopulations
These are the reaction names we need to feed into the visualization.
```{r, results='hide'}
reaction_names = c("Adenosine Phosphorylase", # HMR_8486
                   "Adenosine Kinase", # HMR_4012
                   "Adenine Phosphoribosyltransferase", # HMR_4085
                   "Glutamate Dehydrogenase", # HMR_3802
                   "Inosine Phosphorylase", # HMR_4574
                   "Inosine Kinase", # (INSK)
                   "Hypoxanthine Phosphoribosyltransferase", # HMR_4418
                   "Isocitrate Dehydrogenase (NADPH)",# HMR_4586
                   "Isocitrate Dehydrogenase (NADH)" # HMR_4588
                   )
```

 it onto the HeatMap
```{r, results='hide'}
DoHeatmap(object=neut,
          assay="FBA",
          slot='data',
          label=FALSE,
          draw.lines=TRUE,
          lines.width=10,
          features=subpop_reactions$gene,
          group.colors=c("#4477AA", "#228833", "#AA3377", "#EE6677")) +
          scale_fill_gradient(name="Metabolic flux",
                              low='gray', high='#CC3311') +
          scale_y_discrete(labels=c("HMR-8486"="Adenosine Phosphrylase", 
                                    "HMR-4012"="Adenosine Kinase",
                                    "HMR-4085"="Adenine Phosphoribosyltransferase",
                                    "HMR-3802"="Glutamate Dehydrogenase",
                                    "HMR-4574"="Inosine Phosphorylase",
                                    "INSK"="Inosine Kinase",
                                    "HMR-4418"="Hypoxanthine Phosphoribosyltransferase",
                                    "HMR-4586"="Isocitrate Dehydrogenase (NADPH)",
                                    "HMR-4588"="Isocitrate Dehydrogenase (NADH)")) +
          theme(text=element_text(family='sans', size=16),
                legend.text=element_text(size=16),
                axis.title.x=element_text(size=20),
                axis.title.y=element_text(size=20),
                axis.text.x=element_text(size=16),
                axis.text.y=element_text(size=16),
                legend.title=element_text(size=20)) +
          labs(colour="Neutrophil subtypes")
```