# Modeling the metabolic changes of the epithelial-to-mesenchymal transition 

## Summary
This repository identifies metabolic enzymes that are essential to the epithelial-to-mesenchymal transition (EMT) in the context of lung adenocarcinoma. 

**Three analyses are performed:**

  1. Enrichment and differential expression of multiple lung adenocarcinoma omics datasets (Bulk RNA-Seq, single-cell RNA-Seq, Proteomics, and more)
  2. Constraint-based metabolic reconstruction and analysis for metabolic flux analysis and fitness evaluation from gene and reaction knockouts.
  3. Hypothesis generation from differential flux and growth sensitivty analysis  
  
## Programming languages used in this analysis

  * MATLAB version R2020b Update 4
  * R version 4.03
  * Python version 3.8.6

## Getting Started


## TO DO:
- [ ] Create Docker container for dependencies
- [ ] Clean up all code base further
- [ ] Edit all notebooks

## Usage
COMING SOON
  
## Contributing
Contributions to make this analysis better, more robust, and easier to follow are greatly appreciated. Here are the steps we ask of you:
  1. Fork the project
  2. Create a new branch
  3. Make your changes
  4. Commit your changes
  5. Push to the branch
  6. Open a pull request

## License
Distributed under the GNU License. See `LICENSE` for more information.

## Contact
For questions regarding the code deposited in this repository, please reach out to Scott Campit via email at: scampit [at] umich [dot] edu or via Twitter at @secampit.