---
title: "Processing GSE147405 scRNASeq data for integration with COBRA"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---
  
# Summary
This notebook performs simple exploratory data analysis for single-cell EMT data. 
Some of the code was adapted from various sources.

# 1. Download all libraries and library dependencies
First, we'll download several libraries and library dependencies for single-cell
data processing and data visualization.

Note that the following codeblock takes awhile to run. Only do it if you do not
have the libraries pre-installed in your R environment.
```{r, include=TRUE}
#pkgLst = c("Seurat", "dplyr", "viridis", 
#           "matrixStats", "UpSetR", 
#           "mgcv", "tidyverse", "reshape2", "ggplot2")
#install.packages(pkgLst)

library(mgcv)
library(splines)
library(Seurat)
library(dplyr)
library(tidyr)
library(viridis)
library(matrixStats)
library(UpSetR)
library(reshape2)
library(ggplot2)
```

# 2. Explore the effect of different inducers
The following code block loads the data from Cook and Vanderhyden, 2020 
and plots the data distribution for all genes across all samples. In their
dataset alone, there are 4 cancer cell lines and 3 different inducers:

**Table of cancer cell lines in Cook & Vanderhyden, 2020.**
|CCL             |Tissue       |
|----------------|-------------|
|A549            |Lung         |
|DU145           |Prostate     | 
|MCF7            |Mammary gland|
|OVCA420         |Ovary        |

**EMT Inducers:**
  * $TGF\beta$-1
  * EGF
  * TNF

Due to memory constraints, we're going to have to perform this analysis cancer 
cell line by cancer cell line. First, let's specify some functions to help us.

## Define plotting functions

### Histograms
Next, let's create a function that plots the distribution of the data onto a histogram for each dataset.
```{r, include=TRUE}
plot_histograms = function(data, labels, slot, colors){
  hist_plts = list()
  exp_data = list()

  for (i in 1:length(labels)){
    
    # Extract data
    exp = as.matrix(GetAssayData(data[[i]], slot=slot)) 
    tmp = melt(exp)
    hist_plts[[i]] = ggplot(tmp, aes(x=value)) + 
        
        # Histogram
        geom_histogram(fill=colors[i], bins=100) +
        
        # Density plots
        geom_density(fill=colors[i]) +
        
        # Plot attributes
        xlab("Counts") + ylab("Frequency") + 
        ggtitle(paste(labels[[i]], "UMAP data distribution", sep=" "))
  }
  return(hist_plts)
}
```

### Ridge plots
This function will create Ridge plots for each condition.
```{r, include=TRUE}
plot_ridge = function(data, labels, genes){
  rdgplts = list()
  for (i in 1:length(labels)){
    rdgplts[[i]] = RidgePlot(object=ovca420_tgfb, 
                          features=genes, 
                          ncol=2)
  }
  return(rdgplts)
}
```

### Violin plots
This function will create violin plots for each condition.
```{r, include=TRUE}
plot_violin = function(data, labels, genes){
  vlplt = list()
  for (i in 1:length(labels)){
    vlplt[[i]] = VlnPlot(object=data[[i]], 
                          features=genes)
  }
  return(vlplt)
}
```

### Dot plots
This function will create dot plots for each condition.
```{r, include=TRUE}
plot_dot = function(data, labels, genes){
  dtplt = list()
  for (i in 1:length(labels)){
    dtplt[[i]] = DotPlot(object=data[[i]], 
                          features=genes)
  }
  return(dtplt)
}
```

### UMAP plots
This function will create UMAP scatter plots for each condition.
```{r, include=TRUE}
plot_umap = function(data, labels, genes){
  umpplt = list()
  for (i in 1:length(labels)){
    umpplt[[i]] = FeaturePlot(object=data[[i]], 
                          features=genes)
  }
  return(umpplt)
}
```

### Scree plots
```{r, include=TRUE}
plot_scree = function(obj){
  eigs = obj@reductions$pca@stdev**2
  props = eigs / sum(eigs)
  plot(props, 
       ylab="Proportion of variance", 
       xlab="Principal Component")
}
```

### Heatmaps
This function will create heatmaps for each condition.
```{r, include=TRUE}
plot_hm = function(data, labels, genes){
  hmplt = list()
  for (i in 1:length(labels)){
    hmplt[[i]] = DoHeatmap(object=data[[i]], 
                          features=genes)
  }
  return(hmplt)
}
```

## Inducer Treatment Analysis
For this analysis, I'm trying to see the effects of different inducers on the
data. Let's use the top 20 genes with the highest variance.

### A549
First, let's load up the data from A549.
```{r, include=TRUE}
# Linux
#a549_egf     = readRDS("~/Data/scRNASeq/GSE147405/A549/A549_EGF.rds")
#a549_tgfb    = readRDS("~/Data/scRNASeq/GSE147405/A549/A549_TGFB1.rds")
#a549_tnf     = readRDS("~/Data/scRNASeq/GSE147405/A549/A549_TNF.rds")

# Dell
#a549_egf     = readRDS("D:/Data/scRNASeq/GSE147405/A549/A549_EGF.rds")
#a549_tgfb    = readRDS("D:/Data/scRNASeq/GSE147405/A549/A549_TGFB1.rds")
#a549_tnf     = readRDS("D:/Data/scRNASeq/GSE147405/A549/A549_TNF.rds")

# ACLX
a549_egf     = readRDS("C:/Users/scott/Data/scRNASeq/GSE147405/A549/A549_EGF.rds")
a549_tgfb    = readRDS("C:/Users/scott/Data/scRNASeq/GSE147405/A549/A549_TGFB1.rds")
a549_tnf     = readRDS("C:/Users/scott/Data/scRNASeq/GSE147405/A549/A549_TNF.rds")

a549_labels = c("A549_EGF", "A549_TGFB1", "A549_TNF")
a549_data = c(a549_egf, a549_tgfb, a549_tnf)
```

Note that the data is currently mapped onto HGNC gene symbols rather than Entrez IDs. I need to do that now.
```{r}
#BiocManager::install("biomaRt")
library(biomaRt)
ensembl=useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
```

This function will perform the data mapping and median aggregation.
```{r}
map_hgnc_to_entrez = function(seurat_obj, mart){
  tmp = seurat_obj@assays$RNA@counts
  genes = getBM(
    attributes=c("hgnc_symbol", "entrezgene_id"),
    values=rownames(tmp),
    mart=mart
  )
  
  # Map rows
  mapped_data  = merge(genes, tmp, by.x="hgnc_symbol", by.y=0)
  
  # Remove data points without entrez IDs
  remove_genes = is.na(mapped_data[, "entrezgene_id"])
  mapped_data  = mapped_data[!remove_genes,]
  mapped_data  = mapped_data[, -1]
  
  # Perform median aggregation -> this step takes awhile
  agg_data = aggregate(mapped_data[, -1], 
                       by=list(mapped_data[, 1]), 
                       FUN=median)
  
  # Construct the Assay object to replace in the Seurat obj
  colnames(agg_data)[1] = "entrezgene_id"
  rownames(agg_data) = agg_data[, 1]
  agg_data = agg_data[, -1]
  tmp = CreateAssayObject(counts=agg_data)
  seurat_obj@assays$RNA@counts = tmp@counts
  
  return(seurat_obj)
}

a549_egf = map_hgnc_to_entrez(a549_egf, ensembl)
a549_tgfb = map_hgnc_to_entrez(a549_tgfb, ensembl)
a549_tnf = map_hgnc_to_entrez(a549_tnf, ensembl)
```

Now let's map to the human metabolic reconstruction RECON1


#### Plots of the A549 scRNASeq counts

##### Linear Dimension reduction
First, let's see what the top 20 genes that have the most variance look like. Those will be analyzed downstream.

```{r, include=TRUE}
a549_egf_markers = FindAllMarkers(object=a549_egf,
                         only.pos=TRUE,
                         min.pct=0.25,
                         thresh.use=0.25)
a549_egf_markers = a549_egf_markers[a549_egf_markers$p_val_adj < 0.01, ]

# Get the top 10 for data visualization
a549_egf_markers = a549_egf_markers[1:10, ]

# Run PCA
a549_egf = RunPCA(a549_egf,
              features=VariableFeatures(object=a549_egf))

# Make a Scree Plot
plot_scree(a549_egf)

# View a loadings plot for multiple PCs
VizDimLoadings(a549_egf, 
               dims=1:3, 
               reduction="pca")

# Visualize the scatter plot (2D support only)
DimPlot(a549_egf,
        reduction="pca")

# Visualize the heatmaps
DimHeatmap(a549_egf, 
           dims=1:3, 
           cells=500,
           balanced=TRUE)

```
Then we'll do A549 TGF-B
```{r, include=TRUE}
a549_tgfb_markers = FindAllMarkers(object=a549_tgfb,
                         only.pos=TRUE,
                         min.pct=0.25,
                         thresh.use=0.25)
a549_tgfb_markers = a549_tgfb_markers[a549_tgfb_markers$p_val_adj < 0.01, ]

# Get the top 10
a549_tgfb_markers = a549_tgfb_markers[1:10, ]

# Run PCA
a549_tgfb = RunPCA(a549_tgfb,
                  features=VariableFeatures(object=a549_tgfb))

# Make a Scree Plot
plot_scree(a549_tgfb)

# View a loadings plot for multiple PCs
VizDimLoadings(a549_tgfb, 
               dims=1:3, 
               reduction="pca")

# Visualize the scatter plot (2D support only)
DimPlot(a549_tgfb,
        reduction="pca")

# Visualize the heatmaps
DimHeatmap(a549_tgfb, 
           dims=1:3, 
           cells=500,
           balanced=TRUE)

# Cluster using PCA scores
a549_tgfb = FindClusters(a549_tgfb,
                        reduction.type="pca",
                        dim.use=1:3,
                        resolution=1.2,
                        print.output=0)
```

Finally we'll do A549 TNF
```{r, include=TRUE}
a549_tnf_markers = FindAllMarkers(object=a549_tgfb,
                         only.pos=TRUE,
                         min.pct=0.25,
                         thresh.use=0.25)
a549_tnf_markers = a549_tnf_markers[a549_tnf_markers$p_val_adj < 0.01, ]

# Get the top 10
a549_tnf_markers = a549_tnf_markers[1:10, ]

# Run PCA
a549_tnf = RunPCA(a549_tnf,
              features=VariableFeatures(object=a549_tnf))

# Make a Scree Plot
plot_scree(a549_tnf)

# View a loadings plot for multiple PCs
VizDimLoadings(a549_tnf, 
               dims=1:3, 
               reduction="pca")

# Visualize the scatter plot (2D support only)
DimPlot(a549_tnf,
        reduction="pca")

# Visualize the heatmaps
DimHeatmap(a549_tnf, 
           dims=1:3, 
           cells=500,
           balanced=TRUE)

# Cluster using PCA scores
a549_tnf = FindClusters(a549_tnf,
                        reduction.type="pca",
                        dim.use=1:3,
                        resolution=1.2,
                        print.output=0)
```
We'll create the data structure 
Create data objects containing histograms with raw data.
```{r, include=TRUE}
a549_plt = plot_histograms(a549_data, a549_labels, "data")
for (i in 1:length(a549_labels)){
  print(a549_plt[[i]])
}
```
Create data objects containing histograms with scaled data.
```{r, include=TRUE}
a549_plt = plot_histograms(a549_data, a549_labels, "scale.data")
for (i in 1:length(a549_labels)){
  print(a549_plt[[i]])
}
```

Create data objects containing ridge plots.
```{r, include=TRUE}
a549_plt = plot_ridge(a549_data, a549_labels, a549_egf_markers[, 1])
for (i in 1:length(a549_labels)){
  print(a549_plt[[i]])
}
```

Create data objects containing violin plots.
```{r, include=TRUE}
a549_plt = plot_violin(a549_data, a549_labels, a549_egf_markers[, 1])
for (i in 1:length(a549_labels)){
  print(a549_plt[[i]])
}
```

Create data objects containing UMAP plots.
```{r, include=TRUE}
a549_plt = plot_umap(a549_data, a549_labels, genes)
for (i in 1:length(a549_labels)){
  print(a549_plt[[i]])
}
```

Create data objects containing dot plots.
```{r, include=TRUE}
a549_plt = plot_dot(a549_data, a549_labels, genes)
for (i in 1:length(a549_labels)){
  print(a549_plt[[i]])
}
```

Create data objects containing heatmaps.
```{r, include=TRUE}
a549_plt = plot_hm(a549_data, a549_labels, genes)
for (i in 1:length(a549_labels)){
  print(a549_plt[[i]])
}
```

### DU145
```{r, include=TRUE}
# Linux
#du145_egf    = readRDS("~/Data/scRNASeq/GSE147405/DU145/DU145_EGF.rds")
#du145_tgfb   = readRDS("~/Data/scRNASeq/GSE147405/DU145/DU145_TGFB1.rds")
#du145_tnf    = readRDS("~/Data/scRNASeq/GSE147405/DU145/DU145_TNF.rds")

# Windows
du145_egf    = readRDS("D:/Data/scRNASeq/GSE147405/DU145/DU145_EGF.rds")
du145_tgfb   = readRDS("D:/Data/scRNASeq/GSE147405/DU145/DU145_TGFB1.rds")
du145_tnf    = readRDS("D:/Data/scRNASeq/GSE147405/DU145/DU145_TNF.rds")

du145_labels = c("DU145_EGF", "DU145_TGFB1", "DU145_TNF")
du145_data = c(du145_egf, du145_tgfb, du145_tnf)
```

##### Linear Dimension reduction
First, let's see what the top 10 genes that have the most variance look like. Those will be analyzed downstream.

```{r, include=TRUE}
du145_egf_markers = FindAllMarkers(object=du145_egf,
                         only.pos=TRUE,
                         min.pct=0.25,
                         thresh.use=0.25)
du145_egf_markers = du145_egf_markers[du145_egf_markers$p_val_adj < 0.01, ]

# Get the top 10 for data visualization
du145_egf_markers = du145_egf_markers[1:10, ]

# Run PCA
du145_egf = RunPCA(du145_egf,
              features=VariableFeatures(object=du145_egf))

# Make a Scree Plot
plot_scree(du145_egfdu145_egf)

# View a loadings plot for multiple PCs
VizDimLoadings(du145_egf, 
               dims=1:3, 
               reduction="pca")

# Visualize the scatter plot (2D support only)
DimPlot(du145_egfdu145_egf,
        reduction="pca")

# Visualize the heatmaps
DimHeatmap(du145_egfdu145_egf, 
           dims=1:3, 
           cells=500,
           balanced=TRUE)

```
Then we'll do DU145_egf TGF-B
```{r, include=TRUE}
du145_tgfb_markers = FindAllMarkers(object=du145_tgfb,
                         only.pos=TRUE,
                         min.pct=0.25,
                         thresh.use=0.25)
du145_tgfb_markers = du145_tgfb_markers[du145_tgfb_markers$p_val_adj < 0.01, ]

# Get the top 10
du145_tgfb_markers = du145_tgfb_markers[1:10, ]

# Run PCA
du145_tgfb = RunPCA(du145_tgfb,
              features=VariableFeatures(object=du145_tgfb))

# Make a Scree Plot
plot_scree(du145_tgfb)

# View a loadings plot for multiple PCs
VizDimLoadings(du145_tgfb, 
               dims=1:3, 
               reduction="pca")

# Visualize the scatter plot (2D support only)
DimPlot(du145_tgfb,
        reduction="pca")

# Visualize the heatmaps
DimHeatmap(du145_tgfb, 
           dims=1:3, 
           cells=500,
           balanced=TRUE)

# Cluster using PCA scores
du145_tgfb = FindClusters(du145_tgfb,
                        reduction.type="pca",
                        dim.use=1:3,
                        resolution=1.2,
                        print.output=0)
```

Finally we'll do DU145 TNF
```{r, include=TRUE}
du145_tnf_markers = FindAllMarkers(object=du145_tgfb,
                         only.pos=TRUE,
                         min.pct=0.25,
                         thresh.use=0.25)
du145_tnf_markers = du145_tnf_markers[du145_tnf_markers$p_val_adj < 0.01, ]

# Get the top 10
du145_tnf_markers = du145_tnf_markers[1:10, ]

# Run PCA
du145_tnf = RunPCA(du145_tnf,
              features=VariableFeatures(object=du145_tnf))

# Make a Scree Plot
plot_scree(du145_tnf)

# View a loadings plot for multiple PCs
VizDimLoadings(du145_tnf, 
               dims=1:3, 
               reduction="pca")

# Visualize the scatter plot (2D support only)
DimPlot(du145_tnf,
        reduction="pca")

# Visualize the heatmaps
DimHeatmap(du145_tnf, 
           dims=1:3, 
           cells=500,
           balanced=TRUE)

# Cluster using PCA scores
du145_tnf = FindClusters(du145_tnf,
                        reduction.type="pca",
                        dim.use=1:3,
                        resolution=1.2,
                        print.output=0)
```

Create data objects containing histograms with raw data.
```{r, include=TRUE}
du145_plt = plot_histograms(du145_data, du145_labels, "data")
for (i in 1:length(du145_labels)){
  print(du145_plt[[i]])
}
```
Create data objects containing histograms with scaled data.
```{r, include=TRUE}
du145_plt = plot_histograms(du145_data, du145_labels, "scale.data")
for (i in 1:length(du145_labels)){
  print(du145_plt[[i]])
}
```

Create data objects containing ridge plots.
```{r, include=TRUE}
du145_plt = plot_ridge(du145_data, du145_labels, genes)
for (i in 1:length(du145_labels)){
  print(du145_plt[[i]])
}
```

Create data objects containing violin plots.
```{r, include=TRUE}
du145_plt = plot_violin(du145_data, du145_labels, genes)
for (i in 1:length(du145_labels)){
  print(du145_plt[[i]])
}
```

Create data objects containing UMAP plots.
```{r, include=TRUE}
du145_plt = plot_umap(du145_data, du145_labels, genes)
for (i in 1:length(du145_labels)){
  print(du145_plt[[i]])
}
```

Create data objects containing dot plots.
```{r, include=TRUE}
du145_plt = plot_dot(du145_data, du145_labels, genes)
for (i in 1:length(du145_labels)){
  print(du145_plt[[i]])
}
```

Create data objects containing heatmaps.
```{r, include=TRUE}
du145_plt = plot_hm(du145_data, du145_labels, genes)
for (i in 1:length(du145_labels)){
  print(du145_plt[[i]])
}
```

### MCF7
```{r, include=TRUE}
# Linux
#mcf7_egf     = readRDS("~/Data/scRNASeq/GSE147405/MCF7/MCF7_EGF.rds")
#mcf7_tgfb    = readRDS("~/Data/scRNASeq/GSE147405/MCF7/MCF7_TGFB1.rds")
#mcf7_tnf     = readRDS("~/Data/scRNASeq/GSE147405/MCF7/MCF7_TNF.rds")

# Windows
mcf7_egf     = readRDS("D:/Data/scRNASeq/GSE147405/MCF7/MCF7_EGF.rds")
mcf7_tgfb    = readRDS("D:/Data/scRNASeq/GSE147405/MCF7/MCF7_TGFB1.rds")
mcf7_tnf     = readRDS("D:/Data/scRNASeq/GSE147405/MCF7/MCF7_TNF.rds")

mcf7_labels = c("MCF7_EGF", "MCF7_TGFB1", "MCF7_TNF")
mcf7_data = c(mcf7_egf, mcf7_tgfb, mcf7_tnf)
```

#### Plots of the MCF7 scRNASeq counts

```{r, include=TRUE}
genes = c("SAMD11", "PLEKHN1")
```

Create data objects containing histograms with raw data.
```{r, include=TRUE}
mcf7_plt = plot_histograms(mcf7_data, mcf7_labels, "data")
for (i in 1:length(mcf7_labels)){
  print(mcf7_plt[[i]])
}
```
Create data objects containing histograms with scaled data.
```{r, include=TRUE}
mcf7_plt = plot_histograms(mcf7_data, mcf7_labels, "scale.data")
for (i in 1:length(mcf7_labels)){
  print(mcf7_plt[[i]])
}
```

Create data objects containing ridge plots.
```{r, include=TRUE}
mcf7_plt = plot_ridge(mcf7_data, mcf7_labels, genes)
for (i in 1:length(mcf7_labels)){
  print(mcf7_plt[[i]])
}
```

Create data objects containing violin plots.
```{r, include=TRUE}
mcf7_plt = plot_violin(mcf7_data, mcf7_labels, genes)
for (i in 1:length(mcf7_labels)){
  print(mcf7_plt[[i]])
}
```

Create data objects containing UMAP plots.
```{r, include=TRUE}
mcf7_plt = plot_umap(mcf7_data, mcf7_labels, genes)
for (i in 1:length(mcf7_labels)){
  print(mcf7_plt[[i]])
}
```

Create data objects containing dot plots.
```{r, include=TRUE}
mcf7_plt = plot_dot(mcf7_data, mcf7_labels, genes)
for (i in 1:length(mcf7_labels)){
  print(mcf7_plt[[i]])
}
```

Create data objects containing heatmaps.
```{r, include=TRUE}
mcf7_plt = plot_hm(mcf7_data, mcf7_labels, genes)
for (i in 1:length(mcf7_labels)){
  print(mcf7_plt[[i]])
}
```

### OVCA420
Load the dataset
```{r, include=TRUE}
# Linux
#ovca420_egf  = readRDS("~/Data/scRNASeq/GSE147405/OVCA420/OVCA420_EGF.rds")
#ovca420_tgfb = readRDS("~/Data/scRNASeq/GSE147405/OVCA420/OVCA420_TGFB1.rds")
#ovca420_tnf  = readRDS("~/Data/scRNASeq/GSE147405/OVCA420/OVCA420_TNF.rds")

# Windows
ovca420_egf  = readRDS("D:/Data/scRNASeq/GSE147405/OVCA420/OVCA420_EGF.rds")
ovca420_tgfb = readRDS("D:/Data/scRNASeq/GSE147405/OVCA420/OVCA420_TGFB1.rds")
ovca420_tnf  = readRDS("D:/Data/scRNASeq/GSE147405/OVCA420/OVCA420_TNF.rds")

ovca420_labels = c("OVCA420_EGF", "OVCA420_TGFB1", "OVCA420_TNF")
ovca420_data = c(ovca420_egf, ovca420_tgfb, ovca420_tnf)
```

#### Plots of the OVCA420 scRNASeq counts

```{r, include=TRUE}
genes = c("SAMD11", "PLEKHN1")
```

Create data objects containing histograms with raw data.
```{r, include=TRUE}
ovca420_plt = plot_histograms(ovca420_data, ovca420_labels, "data")
for (i in 1:length(ovca420_labels)){
  print(ovca420_plt[[i]])
}
```
Create data objects containing histograms with scaled data.
```{r, include=TRUE}
ovca420_plt = plot_histograms(ovca420_data, ovca420_labels, "scale.data")
for (i in 1:length(ovca420_labels)){
  print(ovca420_plt[[i]])
}
```

Create data objects containing ridge plots.
```{r, include=TRUE}
ovca420_plt = plot_ridge(ovca420_data, ovca420_labels, genes)
for (i in 1:length(ovca420_labels)){
  print(ovca420_plt[[i]])
}
```

Create data objects containing violin plots.
```{r, include=TRUE}
ovca420_plt = plot_violin(ovca420_data, ovca420_labels, genes)
for (i in 1:length(ovca420_labels)){
  print(ovca420_plt[[i]])
}
```

Create data objects containing UMAP plots.
```{r, include=TRUE}
ovca420_plt = plot_umap(ovca420_data, ovca420_labels, genes)
for (i in 1:length(ovca420_labels)){
  print(ovca420_plt[[i]])
}
```

Create data objects containing dot plots.
```{r, include=TRUE}
ovca420_plt = plot_dot(ovca420_data, ovca420_labels, genes)
for (i in 1:length(ovca420_labels)){
  print(ovca420_plt[[i]])
}
```

Create data objects containing heatmaps.
```{r, include=TRUE}
ovca420_plt = plot_hm(ovca420_data, ovca420_labels, genes)
for (i in 1:length(ovca420_labels)){
  print(ovca420_plt[[i]])
}
```

