---
title: "Computing differentially expressed genes using limma-voom"
author: "Scott Campit"
output: 
  html_notebook: 
    theme: united
    toc: yes
    toc_depth: 2
---

# Summary 
This notebook computes differentially expressed genes using limma-voom for different Epithelial-to-Mesenchymal Transition (EMT) datasets. 

The datasets under consideration are:
  * GSE17708
  * GSE17518
  * Keshamouni et al 2009 (Proteomics)
  * Garcia et al 2019 (Proteomics)

# Differential Gene Expression Analysis (RNASeq)

## Datasets without pre-processed differentially expressed genes
This analysis runs DGEA for GSE17518 and DPEA for the Keshamouni et al 2009 proteomics set.

First, we'll need to load in some libraries.

```{r,warning=FALSE,message=FALSE}
# Install BioCManager and other dependencies to run RNASeq DEG
install.packages(c("BiocManager", "tidyverse", "readxl", 'data.table', 'openxlsx'))
BiocManager::install(c("edgeR", "limma", "GEOquery", "affy",
                       "hgu133plus2.db", "org.Hs.eg.db"), 
                     dependencies=TRUE,
                     update=TRUE)

library(data.table)
library(openxlsx)
library(readxl)

library(hgu133plus2.db)
library(org.Hs.eg.db)
library(edgeR)
library(GEOquery)
library(affy)
library(readxl)
```
### GSE17518

First, we'll need to load in the GSE17518 using GEOQuery. 

```{r,warning=FALSE,message=FALSE}
# Get expression set
tmp = getGEO("GSE17518")
GSE17518 = exprs(tmp[[1]])

# Rename columns
condition = c("0", "0", "0", "72", "72", "72")
colnames(GSE17518) = condition
```

Calculating normalized data from `DGEList` object.

```{r}
# Calculate the normalization factors
d0 = DGEList(abs(GSE17518))
d0 = calcNormFactors(d0)
```

Filter lowly expressed genes using counts per million (CPM)

```{r}
# Filter using CPM for lowly expressed genes.
cutoff = 0.5
drop   = which(apply(cpm(d0), 1, max) < cutoff)
d      = d0[-drop,] 
dim(d) 
```

Multidimensional scaling plot 

```{r}
# Multidimensional scaling plot
plotMDS(d, col=as.numeric(condition))
```

Now, let's transform the dataset using *voom*.
```{r}
# Voom transformation 
mm <- model.matrix(~0 + condition)
y = voom(d, mm, plot=T)
```

Now we can fit the data using *limma*

```{r}
fit <- lmFit(y, mm)
head(coef(fit))
```

We can make comparisons between the two groups. This is equivalent to the $log_{2}$ fold change.
```{r}
contr <- makeContrasts(condition72 - condition0, levels = colnames(coef(fit)))
contr
```

Now let's estimate the contrast for each gene.

```{r}
tmp = contrasts.fit(fit, contr)
```

We'll now use Empirical Bayes to smooth out the standard errors.

```{r}
tmp <- eBayes(tmp)
```

We can now look at differentially expressed genes in the `top.table` variable.

```{r}
top.table <- topTable(tmp, sort.by = "P", n = Inf)
head(top.table, 20)

# View the number of DE genes
length(which(top.table$adj.P.Val < 0.05)) 
```

Map the identifiers to symbols and Entrez

```{r}
# Format DGE table
df = setDT(top.table)
probes = as.character(rownames(top.table))

# Get identifiers
map = select(hgu133plus2.db, 
               keys=probes, 
               columns = c('PROBEID', 
                           'ENTREZID', 
                           'SYMBOL')
               )

# Join map and data
df = cbind(probes, df)
mapped_lst = merge(map, df, by.x="PROBEID", by.y="probes")

# Save to file
filename = '~/Data/RNASeq/GSE17518/GSE17518_DGE.xlsx'
write.xlsx(mapped_lst, filename)
```

# Differential Protein Expression analysis (LC/MS-MS)

First, we'll need to load in the GSE17518 using GEOQuery. 

```{r,warning=FALSE,message=FALSE}
# Get expression set
filename = '~/Data/Proteomics/Keshamouni/GSE17518_series_matrix_IMR90FibroblastTGFb.xlsx'
tmp = read_excel(filename)

# Rename columns
condition = c("0", "0", "0", "72", "72", "72")
colnames(GSE17518) = condition
```

Calculating normalized data from `DGEList` object.

```{r}
# Calculate the normalization factors
d0 = DGEList(abs(GSE17518))
d0 = calcNormFactors(d0)
```

Filter lowly expressed genes using counts per million (CPM)

```{r}
# Filter using CPM for lowly expressed genes.
cutoff = 0.5
drop   = which(apply(cpm(d0), 1, max) < cutoff)
d      = d0[-drop,] 
dim(d) 
```

Multidimensional scaling plot 

```{r}
# Multidimensional scaling plot
plotMDS(d, col=as.numeric(condition))
```

Now, let's transform the dataset using *voom*.
```{r}
# Voom transformation 
mm <- model.matrix(~0 + condition)
y = voom(d, mm, plot=T)
```

Now we can fit the data using *limma*

```{r}
fit <- lmFit(y, mm)
head(coef(fit))
```

We can make comparisons between the two groups. This is equivalent to the $log_{2}$ fold change.
```{r}
contr <- makeContrasts(condition72 - condition0, levels = colnames(coef(fit)))
contr
```

Now let's estimate the contrast for each gene.

```{r}
tmp = contrasts.fit(fit, contr)
```

We'll now use Empirical Bayes to smooth out the standard errors.

```{r}
tmp <- eBayes(tmp)
```

We can now look at differentially expressed genes in the `top.table` variable.

```{r}
top.table <- topTable(tmp, sort.by = "P", n = Inf)
head(top.table, 20)

# View the number of DE genes
length(which(top.table$adj.P.Val < 0.05)) 
```

Map the identifiers to symbols and Entrez

```{r}
# Format DGE table
df = setDT(top.table)
probes = as.character(rownames(top.table))

# Get identifiers
map = select(hgu133plus2.db, 
               keys=probes, 
               columns = c('PROBEID', 
                           'ENTREZID', 
                           'SYMBOL')
               )

# Join map and data
df = cbind(probes, df)
mapped_lst = merge(map, df, by.x="PROBEID", by.y="probes")

# Save to file
filename = '~/Data/RNASeq/GSE17518/GSE17518_DGE.xlsx'
write.xlsx(mapped_lst, filename)
```