---
title: "Computing differentially expressed genes using limma-voom"
author: "Scott Campit"
output: 
  html_notebook: 
    theme: united
    toc: yes
    toc_depth: 2
---

# Summary 
This notebook computes differentially expressed genes using limma-voom for different Epithelial-to-Mesenchymal Transition (EMT) datasets. 

The datasets under consideration are:
  * GSE17708
  * GSE17518
  * Keshamouni et al 2009 (Proteomics)
  * Garcia et al 2019 (Proteomics)

# Differential Gene Expression Analysis (RNASeq)

First, we'll need to load in some libraries.

```{r,warning=FALSE,message=FALSE}
# Install BioCManager and other dependencies to run RNASeq DEG
install.packages(c("BiocManager", "tidyverse", "readxl"))
BiocManager::install(c("edgeR", "limma", "GEOquery", "affy"), 
                     dependencies=TRUE,
                     update=TRUE)

library(edgeR)
library(GEOquery)
library(affy)
library(readxl)
```
# Get expression set from GSE17708 and GSE17518
Next, we'll need to load in the RNASeq datasets using GEOQuery. 
Note that the GSE identifiers are mapped to the following studies:
  * GSE17708: Keshamouni et al., 2009
    * Notes: 
      * There are 9 time points, and each experiment was repeated 3 times
      * Samples were assayed using Affymetrix HG_U133_plus_2 arrays
      
  * GSE17518: Thannickal et al., 2009

```{r,warning=FALSE,message=FALSE}
# GSE17708 -> raw files don't have probe id, so what I ended up doing was 
# undoing the log transform in the mapped dataset.
#tarArchive  = "/home/scampit/Software/emt/src/r/GSE17708/GSE17708_RAW/"
#cels        = list.files(tarArchive, pattern="[.gz]")
#celpath     = sapply(paste(tarArchive, cels, sep="/"))
#data        = ReadAffy(filenames=celpath)
#GSE17708    = exprs(data)
#logGSE17708 = log2(GSE17708) 

filename = "/home/scampit/Software/emt/src/r/GSE17708/GSE17708_Keshamouni_TGFB1_logs.xls"
data = read_excel(filename, sheet="LogData", skip=4)
tmp = data[, -c(1:5)]
tmp = 10^tmp

install.packages(c("plotly", "ggplot2", "tidyr"))
library(plotly)
library(ggplot2)
library(tidyr)

ggplot(data=gather(tmp), aes(value)) + geom_histogram()


# GSE17518
tmp = getGEO("GSE17518")
GSE17518 = tmp[[1]]
```

# Differential Protein Expression analysis (LC/MS-MS)

