---
title: "subpopulatoin analysis"
output: html_notebook
---

# Summary
This notebook performs simple exploratory data analysis for single-cell EMT data. 
Some of the code was adapted from various sources.

# 1. Download all libraries and library dependencies
First, we'll download several libraries and library dependencies for single-cell
data processing and data visualization.

Note that the following codeblock takes awhile to run. Only do it if you do not
have the libraries pre-installed in your R environment.
```{r, include=TRUE}
#pkgLst = c("Seurat", "dplyr", "viridis", 
#           "matrixStats", "UpSetR", 
#           "mgcv", "tidyverse", "reshape2", "ggplot2")
#install.packages(pkgLst)

library(mgcv)
library(splines)
library(Seurat)
library(dplyr)
library(tidyr)
library(viridis)
library(matrixStats)
library(UpSetR)
library(reshape2)
library(ggplot2)
library(Rmagic)

```

# 2. Load A549 TGF-B dataset
First, let's load up the data from A549.
```{r, include=TRUE}
rm(list=ls())
#datapath = "C:/Users/scott/Data/scRNASeq/GSE147405/A549/A549_TGFB1.rds"
datapath = "D:/Data/scRNASeq/GSE147405/A549/A549_TGFB1.rds"
a549_tgfb = readRDS(datapath)
unique(a549_tgfb@meta.data$Time)
```
## Filter genes
Let's filter out some low quality genes.
```{r}
genes_filter = rownames(a549_tgfb[["RNA"]]@data)[rowSums(as.matrix(a549_tgfb[["RNA"]]@data)) > 0]
a549_tgfb = subset(a549_tgfb, features=genes_filter)
```

## Regress out cell cycle
We can regress some biological variance out, including cell cycle genes.
```{r}
genes_to_regress = "D:/Data/scRNASeq/GSE147405/regev_lab_cell_cycle_genes.txt"
cc.genes = readLines(con=genes_to_regress)
s.genes  = cc.genes[1:43]
g2m.genes = cc.genes[44:97]
a549_tgfb = CellCycleScoring(a549_tgfb, 
                             s.features=s.genes,
                             g2m.features=g2m.genes, 
                             set.ident=FALSE)
```

## Subset time points of interest
Finally, let's subset out some of the cells.
```{r}
a549_tgfb = subset(a549_tgfb, subset = Time==c("0d", "8h", "3d", "1d", "7d"))
```

## Preprocess data for MAGIC
Now, we'll run MAGIC on the dataset.
```{r}
data = data.frame(a549_tgfb@assays$RNA@counts)
data = t(data)
```

Ensure our initial filtering worked.
```{R}
# keep genes expressed in at least 10 cells
keep_cols = colSums(data > 0) > 10
data = data[,keep_cols]
```

Plot the data distribution.
```{r}
# look at the distribution of library sizes
ggplot() +
  geom_histogram(aes(x=rowSums(data)), bins=50) +
  geom_vline(xintercept = 1000, color='red')
```

Normalize by library size and take square root.
```{r}
data = library.size.normalize(data)
data = sqrt(data)
```

After this, all we have to do is run the dataset through MAGIC in Python.
```{r}
write.table(data, file="D:/Analysis/EMT/a549_tgfb_preprocessed_for_magic.csv", 
            row.names=TRUE, col.names=TRUE)
```

# 3. Data preprocessing
```{r, results='hide'}
find_subpopulations = function(gene_obj){
  
  # Identify variable features
  gene_obj = FindVariableFeatures(object=gene_obj, 
                              mean.function=ExpMean,
                              dispersion.function=LogVMR,
                              selection.method='vst',
                              x.low.cutoff=-Inf,
                              x.high.cutoff=Inf,
                              y.cutoff=Inf,
                              nfeatures = 2000)
  
  gene_obj <- SCTransform(gene_obj,
                         do.correct.umi=T,
                         vars.to.regress=c("S.Score", "G2M.Score",
                                           "percent.mito"),
                         #batch_var="Mix",
                         return_gene_attr=T,
                         variable.features.n=2000)

  
  # Run PCA
  gene_obj = RunPCA(object=gene_obj,
                verbose=FALSE)
  
  # Find shortest nearest neighbors and clusters
  gene_obj = FindNeighbors(gene_obj, 
                       dims=1:30,
                       reduction='pca')
  gene_obj = FindClusters(gene_obj, 
                      resolution=0.5)
  
  # Perform T-SNE
  gene_obj = RunTSNE(gene_obj,
                     dims=1:30)
  
  # Perform UMAP
  gene_obj = RunUMAP(gene_obj, 
                 dims=1:30)
  
  return(gene_obj)
}
```

```{r}
a549_tgfb = find_subpopulations(a549_tgfb)
cluster_pred = a549_tgfb@active.ident
```

```{r}
a549_tgfb@active.ident = as.factor(a549_tgfb@meta.data$Time)
DimPlot(a549_tgfb, 
        reduction="umap",
        dims=c(1, 2),
        #cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="", 
             x="UMAP 1", 
             y="UMAP 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.15),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))

a549_tgfb@active.ident = cluster_pred
DimPlot(a549_tgfb, 
        reduction="umap",
        dims=c(1, 2),
        #cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="", 
             x="UMAP 1", 
             y="UMAP 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.15),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))
```