---
title: "CRISPR-Cas9 Validation"
author: "Scott Campit"
output: html_notebook
---

# Summary
This notebook does the following analysis:
  1. Separate out lung cancer cell lines
  2. Compute differentially expressed CS
  3. Compare it to current iteration of metabolic reactions and genes predicted by COBRA

# 1. Load libraries
First, let's load some essential libraries.
```{r}
# Essential libraries
library(tidyr)
library(readr)
library(stringr)
library(ggplot2)
library(reticulate)
library(tidyverse)
library(dplyr)
library(reshape2)
library(readxl)
library(data.table)

# Bioinformatics libraries
library(limma)
library(Glimma)
library(edgeR)

# Data visualization libraries
library(ggVennDiagram)
library(stats)
library(ggpubr)
require(gridExtra)
library(extrafont)
library(ComplexHeatmap)
library(circlize)

# Statistics / DS libraries
library(rrcov)
```

# 2. Read in CRISPR Essentiality Data
There are two datasets we can read in:
  A. Robust PCA dataset performed on CERES Scores
  B. Batch normalized CERES Scores

Either way, the downstream analyses will be the same.

## A. Load the Robust PCA CERES Scores
This data was filtered based on outliers: data points that contain high leverage and orthogonal distance. This analysis used the `rrcov` library. 

More technical information about this analysis can be found in this paper: Hubert, M., Rousseeuw, P. J., & Vanden Branden, K. (2005). ROBPCA: a new approach to robust principal component analysis. Technometrics, 47(1), 64-79.
```{r}
rm(list=ls())
datapath = "D:/Analysis/EMT/data/ceres_robpca.RData"
load(datapath)
```

## B. Load the Batch Normalized CERES Scores
This data was obtained from Pacini, C., Dempster, J. M., Boyle, I., Gon√ßalves, E., Najgebauer, H., Karakoc, E., ... & Iorio, F. (2021). Integrated cross-study datasets of genetic dependencies in cancer. Nature communications, 12(1), 1-14. They combined the CERES Scores from the Broad and Sanger Institute after batch normalization. 

```{r}
rm(list=ls())
datapath = "D:/Analysis/EMT/data/batched_ceres.RData"
load(datapath)
```

# 3. Compute P-values from the CERES Score data distribution
To compute the P-values, there are two ways we can look at the data:
  A. Given the entire dataset, we can see how far the data is from the mean (e.g., a Z-score like metric)
  B. Given two conditions (Primary verus. Metastatic), we can compare the two conditions.
  
## A. Compute P-value across entire data distribution
Next, we'll compute the p-values using the data itself. Most of the data lies close to 0 (no activity), while the significant data points lie on the left side of the data distribution. Thus, we'll compute the left-tailed P-value.

### i. Compute left-tailed p-value
This codeblock computes the left tailed p-value. I also tried computing an FDR adjusted p-value, but none of the genes turn out to be significant.
```{r}
x       = match_data
mu      = mean(x, na.rm=TRUE)
stdev   = sd(x, na.rm=TRUE)
pvalmat = apply(x, 2, function(x) pnorm(x, 
                                        mean=mu,
                                        sd=stdev,
                                        lower.tail=TRUE))
padjmat = apply(pvalmat, 2, function(x) p.adjust(x, method="fdr"))
sum(padjmat < 0.05, na.rm=TRUE)
```

### ii. SANITY CHECK: Plot the data distribution
First, we'll preprocess data for a histogram
```{r}
mlt_pval = reshape2::melt(padjmat)
mlt_pval$col = as.factor(mlt_pval$value < 0.05)
mlt_score = reshape2::melt(x)
mlt_score$col = as.factor(mlt_pval$value < 0.05)
```

Now let's plot the data distribution. Note that we'll be plotting the FDR, which has ~ 325413 significant data points in total.
```{r, fig.height=7.5, fig.width=10, dpi=100}
# Plot p-value distribution
p1 = ggplot(mlt_pval, aes(x=value, fill=col)) +  
  geom_histogram(bins=100) +
  scale_fill_manual(values = c("TRUE"="red", "FALSE"="gray")) +
  labs(title="CERES FDR distribution (all)", 
       x="FDR Adjusted P-values (left-sided; alpha=0.05)", 
       y="Frequency") +
  theme(legend.position="none")

# Plot CERES distribution
p2 = ggplot(mlt_score, aes(x=value, fill=col)) +  
  geom_histogram(bins=100) +
  scale_fill_manual(values = c("TRUE"="red", "FALSE"="gray")) +
  labs(title="CERES distribution (Batch corrected)", 
       x="CERES Score (Batch Corrected)", 
       y="Frequency") +
  theme(legend.position="none")

grid.arrange(p1, p2)
```

# 4. Grab NSCLC Lung Cancer Cell Line Data
The `cellmatch` dataframe contains the metadata for the `match_data`. We're interested in specifically analyzing NSCLC data.

## A. Subset the Data
```{r}
NSCLC_meta = cellmatch[cellmatch["lineage_subtype"]=="NSCLC", ]
NSCLC_meta = NSCLC_meta[!is.na(NSCLC_meta$primary_or_metastasis), ]
``` 

## B. Visualize Primary V. Metastatic
Let's visualize how many cells are primary site versus metastatic cancers. First, let's get the counts.
```{r}
unique(NSCLC_meta$primary_or_metastasis)
tmp = NSCLC_meta %>%
  group_by(primary_or_metastasis) %>%
  count()
```

Now let's make the barplot.
```{r, fig.width=5, fig.height=5, dpi=100}
p3 = ggplot(data=tmp, aes(x=primary_or_metastasis, 
                          y=n, 
                          fill=primary_or_metastasis)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("Metastasis"="#EF476F", "Primary"="#073B4C")) +
  labs(title="NSCLC cell line distribution", 
       x="Type of NSCLC cell line", 
       y="Frequency") +
  theme(legend.position="none")
p3
```
While the data is not skewed that much, it is still notable that there are more metastatic versus primary site NSCLC cell lines in the data.

## C. Grab NSCLC specific data
Finally, we'll grab the CERES scores and the FDR associated with each NSCLC cell line.
```{r}
NSCLC_ceres = match_data[, colnames(match_data) %in% NSCLC_meta$DepMap_ID]
NSCLC_fdr = padjmat[,  colnames(padjmat) %in% NSCLC_meta$DepMap_ID]
```

## D. Grab RECON1 metabolic reconstruction genes in the NSCLC dataset
To benchmark RECON1 results, we'll grab the CERES score gene entries that intersect with the metabolic reconstruction.
```{r}
# Grab map between Entrez and Symbol in RECON1
ids = rownames(NSCLC_ceres)
idx = ids %in% recon1_genes$Symbol

recon1_nsclc_ceres = NSCLC_ceres[idx, ]
recon1_nsclc_fdr = NSCLC_fdr[idx, ]
```

## E. Visualize the data
Now that we have the entire subset, let's visualize the data by splitting the data between primary and metastatic cell lines. 

First, we'll separate the data into metastatic and primary.
```{r}
met_idx = NSCLC_meta$primary_or_metastasis == "Metastasis"
prim_idx = NSCLC_meta$primary_or_metastasis == "Primary"
```

Then we'll pre-process some matrices for visualization via heatmap.
```{r}
# Metastatic data
met_data = recon1_nsclc_ceres[, met_idx]
met_fdr = recon1_nsclc_fdr[, met_idx]
met_bin = met_fdr > 0.05

# Turn non-significant datapoints into black
masked_met_data = met_data
masked_met_data[met_bin] = NA

# Primary site data
prim_data = recon1_nsclc_ceres[, prim_idx]
prim_fdr = recon1_nsclc_fdr[, prim_idx]
prim_bin = prim_fdr > 0.05

# Turn non-significant datapoints into black
masked_prim_data = prim_data
masked_prim_data[prim_bin] = NA
```

### i. Create annotation bar for primary/metastatic
```{r}
h1_col = HeatmapAnnotation(hm1=anno_block(gp=gpar(fill="#EF476F"),
                                      labels=c("Metastatic"), 
                                      labels_gp=gpar(col="black", 
                                                     fontsize=10)))

h2_col = HeatmapAnnotation(hm2=anno_block(gp=gpar(fill="#073B4C"),
                                      labels=c("Primary"), 
                                      labels_gp=gpar(col="white", 
                                                     fontsize=10)))
```

### ii. Visualize all data
```{r, fig.width=12.5, fig.height=40, fig.align='right', dpi=100}
# Metastatic heatmap
hm1 = Heatmap(as.matrix(met_data),
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=TRUE,
              bottom_annotation=h1_col, 
              #width = ncol(met_data)*unit(3, "mm"), 
              #height = nrow(met_data)*unit(3, "mm"),
              show_heatmap_legend=FALSE)
idx = row_order(hm1)



# Primary heatmap
hm2 = Heatmap(as.matrix(prim_data)[idx, ],
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              bottom_annotation=h2_col, 
              #width = ncol(prim_data)*unit(3, "mm"), 
              #height = nrow(prim_data)*unit(3, "mm"),
              heatmap_legend_param=list(title="CERES Score"))

all_nsclc = hm1 + hm2
all_nsclc
```

### iii. Visualize the signficant data only
```{r, fig.width=12.5, fig.height=40, fig.align='right', dpi=100}
# Metastatic heatmap
col_fun = colorRamp2(c(0, median(masked_met_data, na.rm=TRUE), min(masked_met_data, na.rm=TRUE)), 
                     c("#FFFFF9", "#FFCE03", "#FF0000"))
hm3 = Heatmap(as.matrix(masked_met_data)[idx, ],
              na_col = "black",
              col=col_fun,
              cluster_columns=FALSE,
              cluster_rows=FALSE,
              bottom_annotation=h1_col, 
              #width = ncol(met_data)*unit(3, "mm"), 
              #height = nrow(met_data)*unit(3, "mm"),
              show_heatmap_legend=FALSE)

# Primary heatmap
hm4 = Heatmap(as.matrix(masked_prim_data)[idx, ],
              na_col = "black",
              col=col_fun,
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              bottom_annotation=h2_col, 
              #width = ncol(prim_data)*unit(3, "mm"), 
              #height = nrow(prim_data)*unit(3, "mm"),
              heatmap_legend_param=list(title="CERES Score"))

significant_nsclc_only = hm3 + hm4
significant_nsclc_only
```

### iv. Visualize the bottom 50 metabolic genes
```{r}
tmp1 = as.matrix(masked_met_data)[idx, ]
tmp2 = as.matrix(masked_prim_data)[idx, ]

# Metastatic heatmap
hm5 = Heatmap(as.matrix(tmp1),
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=TRUE,
              bottom_annotation=h1_col, 
              #width = ncol(met_data)*unit(3, "mm"), 
              #height = nrow(met_data)*unit(3, "mm"),
              show_heatmap_legend=FALSE)



# Primary heatmap
hm6 = Heatmap(as.matrix(tmp2,
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              bottom_annotation=h2_col, 
              #width = ncol(prim_data)*unit(3, "mm"), 
              #height = nrow(prim_data)*unit(3, "mm"),
              heatmap_legend_param=list(title="CERES Score"))

all_nsclc = hm1 + hm2
all_nsclc

```


Now actually plot the p-values and CERES scores.
  * Note this doesn't look right.
```{r}
# Plot p-value distribution
ggplot(data.frame(tmp), aes(x=value, fill=col)) +  
  geom_histogram(bins=100) +
  scale_fill_manual(values = c("TRUE"="red", "FALSE"="gray")) +
  labs(title="CERES NSCLC p-value distribution", 
       x="P-values (left-sided)", 
       y="Frequency") +
  theme(legend.position="none")

# Plot CERES distribution
ggplot(tmp, aes(x=ceres, fill=col)) +  
  geom_histogram(bins=100) +
  scale_fill_manual(values = c("TRUE"="red", "FALSE"="gray")) +
  labs(title="CERES NSCLC distribution", 
       x="CERES", 
       y="Frequency") +
  theme(legend.position="none")
```
# 6. A549-specific p-values
```{r}
a549_ceres = merged_lung[merged_lung$stripped_cell_line_name=="A549",]
tmp = as.numeric(a549_ceres[, 6:ncol(a549_ceres)])
mu = mean(tmp, na.rm=TRUE)
stdev = sd(tmp, na.rm=TRUE)
a549_pvals = pnorm(tmp, 
                     mean=mu, 
                     sd=stdev,
                     lower.tail=TRUE)
a549_data = data.frame(tmp)
a549_data$pval = a549_pvals
a549_data$entrez = colnames(a549_ceres[, 6:ncol(a549_ceres)]) 


xname = data.frame(entrez=colnames(a549_ceres[, 6:ncol(a549_ceres)]))
tmp = merge(xname, recon1_match, by='entrez')
tmp[, 3:4] = NULL
tmp = unique(tmp)
a549_data$Symbol = tmp$symbol
sig_a549_ceres = a549_ceres[a549_ceres$PValue < 0.05, ]
```


# 7. Get differentially expressed CERES score by comparing metastatic / primary site using limma-voom

## Data preprocessing
Format the dataset
```{r}
sample_info = merged_lung[, 1:5]
data = merged_lung[, -c(1:5)]
rownames(data) = sample_info[, 2]
data = t(data)
sample_info = sample_info %>% dplyr::select(2:3)
```

Now we'll floor positive CERES Scores to 0, and take the absolute value for the CERES Scores
```{r}
data[data > 0] = 0
data = abs(data)
data[is.na(data)] = 0
```

## Preparing limma-voom data 
Now let's preprocess the data
```{r}
# Create a DGE List data object
data = cpm(data)
d0 = DGEList(data)
d0 = calcNormFactors(d0)

# Filter out bad data points
cutoff = 1
drop = which(apply(cpm(d0), 1, max) < cutoff)
d = d0[-drop,] 
dim(d)
```

We need to set up the groups we want to compare and the design matrix.
```{r}
group = as.factor(sample_info$primary_or_metastasis)
design = model.matrix(~0+group)
```

## Run limma-voom
Now let's run *limma-voom*
```{r}
v = voom(d, design, plot=TRUE)
fit = lmFit(v, design, method='robust')
result = eBayes(fit, robust=TRUE)
```

Finally, let's get the top 20 results
```{r}
top.table = topTable(result)
head(top.table, 20)
write.table(top.table, "clipboard", sep="\t", col.names=TRUE)
```

# 8. Get gene list that are universally lethal across all cell lines
This gene list was determined from Project Score. Specifically, I defined universally lethal genes that are lethal in more than 50% of the cancer cell lines out of > 300. 
```{r}
project_score_file = "D:/Data/CRISPR/Score/41586_2019_1103_MOESM4_ESM/Supplementary Table 3 - Core-Fitness and Context-Specificity Summary_v2.xlsx"
sheet_name = "Filter Genes"

project_score = read_excel(project_score_file, sheet_name)
genes_to_filter = project_score[project_score$`Dependency Score` > 50, ]
```

# 9. Get differentially expressed genes for A549 from CERES score
First, we'll get the row corresponding to the cancer cell line A549.
```{r}
a549_ceres = merged_lung[merged_lung$stripped_cell_line_name=="A549",]
a549_ceres[2, ] = merged_pval[merged_pval$stripped_cell_line_name=="A549",]
a549_ceres = a549_ceres[, 6:ncol(a549_ceres)]
a549_ceres = t(a549_ceres)
a549_ceres = data.frame(a549_ceres)
colnames(a549_ceres) = c("CERES", "PValue")
```

Now, let's grab the metabolic genes that are in RECON1
```{r}
xname = data.frame(entrez=rownames(a549_ceres))
tmp = merge(xname, recon1_match, by='entrez')
tmp[, 3:4] = NULL
tmp = unique(tmp)
a549_ceres$Symbol = tmp$symbol
```

## Plot Venn Diagram
```{r}
gene_overlap = list(
  A549_specific = sig_a549_ceres$Symbol,
  Core_fitness = genes_to_filter$`HGNC Symbol`
)
ggVennDiagram(gene_overlap)
```

Finally, we'll grab the genes that have a left-tailed p-value less than 0.05
```{r}
sig_a549_ceres = a549_ceres[a549_ceres$PValue < 0.05, ]
genes_to_filter = genes_to_filter$`Entrez ID`
genes_to_filter = data.frame(genes_to_filter)
ubiq_genes = merge(genes_to_filter, 
                   sig_a549_ceres, 
                   by.x='genes_to_filter', 
                   by.y="row.names")
unique_genes = sig_a549_ceres[!sig_a549_ceres$Symbol %in% ubiq_genes$Symbol, ]
write.table(sig_a549_ceres, "clipboard", sep="\t", col.names=TRUE)
```

# 10. Identify where the predicted knockouts are in the CERES distribution
First, let's load the model predictions
```{r}
recon1_prediction = "D:/Analysis/EMT/EMT_RECON1_KO_PRED.xlsx"
pred = read_excel(recon1_prediction)
a549_ceres$is_pred = rownames(a549_ceres) == pred$Entrez
a549_ceres$category = "A549"
sum(a549_ceres$is_pred)
```

Now let's create the histogram
```{r}
# Create histogram of reaction distribution
p1 = ggplot(a549_ceres, aes(x=CERES, fill=is_pred)) +
  geom_histogram(bins=100, show.legend=TRUE) +
  theme_minimal() +
  theme(legend.position="none") +
  xlim(-0.5, 0.5) +
  theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_blank(),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin()) +
  labs(title="",
       x=element_blank(),
       y="Frequency") + 
     scale_color_manual(values=c("TRUE" = "red", 
                                 "FALSE" = "gray"),
                        aesthetics = c("colour", "fill"))
```

Let's create the scatter plot
```{r}
p2 = ggplot(a549_ceres, aes(x=CERES, y=category, color=is_pred)) +
     geom_point(show.legend=FALSE, alpha=0.01) +
     geom_jitter() +
     theme_minimal() + 
     theme(legend.position="none") +
     xlim(-0.5, 0.5) +
     labs(x="CERES Score",
          y=element_blank()) + 
     scale_color_manual(values=c("TRUE" = "red", 
                                 "FALSE" = "gray"),
                        aesthetics = c("colour", "fill"))
```

Finally construct the figure.
```{r}
figure = ggarrange(p1, p2,
                   ncol = 1, nrow = 2)
figure
```