---
title: "CRISPR-Cas9 Validation"
author: "Scott Campit"
output: html_notebook
---

# Summary
This notebook does the following analysis:
  1. Separate out lung cancer cell lines
  2. Compute differentially expressed CS
  3. Compare it to current iteration of metabolic reactions and genes predicted by COBRA

# 1. Load libraries
First, let's load some essential libraries.
```{r}
rm(list=ls())
# Essential libraries
library(tidyr)
library(readr)
library(stringr)
library(ggplot2)
library(reticulate)
library(tidyverse)
library(dplyr)
library(reshape2)
library(readxl)
library(data.table)

# Bioinformatics libraries
library(limma)
library(Glimma)
library(edgeR)

# Data visualization libraries
library(ggVennDiagram)
library(stats)
library(ggpubr)
require(gridExtra)
library(extrafont)
library(ComplexHeatmap)
library(circlize)

# Statistics / DS libraries
library(rrcov)
```

# 2. Read in CRISPR Essentiality Data
There are two datasets we can read in:
  A. Robust PCA dataset performed on CERES Scores
  B. Batch normalized CERES Scores

Either way, the downstream analyses will be the same.

## A. Load the Robust PCA CERES Scores
This data was filtered based on outliers: data points that contain high leverage and orthogonal distance. This analysis used the `rrcov` library. 

More technical information about this analysis can be found in this paper: Hubert, M., Rousseeuw, P. J., & Vanden Branden, K. (2005). ROBPCA: a new approach to robust principal component analysis. Technometrics, 47(1), 64-79.
```{r}
rm(list=ls())
datapath = "D:/Analysis/EMT/data/ceres_robpca.RData"
load(datapath)
```

## B. Load the Batch Normalized CERES Scores
This data was obtained from Pacini, C., Dempster, J. M., Boyle, I., Gonçalves, E., Najgebauer, H., Karakoc, E., ... & Iorio, F. (2021). Integrated cross-study datasets of genetic dependencies in cancer. Nature communications, 12(1), 1-14. They combined the CERES Scores from the Broad and Sanger Institute after batch normalization. 

```{r}
rm(list=ls())
datapath = "D:/Analysis/EMT/data/batched_ceres.RData"
load(datapath)
```

# 3. Compute P-values from the CERES Score data distribution
To compute the P-values, there are two ways we can look at the data:
  A. Given the entire dataset, we can see how far the data is from the mean (e.g., a Z-score like metric)
  B. Given two conditions (Primary verus. Metastatic), we can compare the two conditions.
  
## A. Compute P-value across entire data distribution
Next, we'll compute the p-values using the data itself. Most of the data lies close to 0 (no activity), while the significant data points lie on the left side of the data distribution. Thus, we'll compute the left-tailed P-value.

### i. Compute left-tailed p-value
This codeblock computes the left tailed p-value. I also tried computing an FDR adjusted p-value, but none of the genes turn out to be significant.
```{r}
x       = match_data
mu      = mean(x, na.rm=TRUE)
stdev   = sd(x, na.rm=TRUE)
pvalmat = apply(x, 2, function(x) pnorm(x, 
                                        mean=mu,
                                        sd=stdev,
                                        lower.tail=TRUE))
padjmat = apply(pvalmat, 2, function(x) p.adjust(x, method="fdr"))
sum(padjmat < 0.05, na.rm=TRUE)
```

### ii. SANITY CHECK: Plot the data distribution
First, we'll preprocess data for a histogram
```{r}
mlt_pval = reshape2::melt(padjmat)
mlt_pval$col = as.factor(mlt_pval$value < 0.05)
mlt_score = reshape2::melt(x)
mlt_score$col = as.factor(mlt_pval$value < 0.05)
```

Now let's plot the data distribution. Note that we'll be plotting the FDR, which has ~ 325413 significant data points in total.
```{r, fig.height=7.5, fig.width=10, dpi=100}
# Plot p-value distribution
p1 = ggplot(mlt_pval, aes(x=value, fill=col)) +  
  geom_histogram(bins=100) +
  scale_fill_manual(values = c("TRUE"="red", "FALSE"="gray")) +
  labs(title="CERES FDR distribution (all)", 
       x="FDR Adjusted P-values (left-sided; alpha=0.05)", 
       y="Frequency") +
  theme(legend.position="none")

# Plot CERES distribution
p2 = ggplot(mlt_score, aes(x=value, fill=col)) +  
  geom_histogram(bins=100) +
  scale_fill_manual(values = c("TRUE"="red", "FALSE"="gray")) +
  labs(title="CERES distribution (Batch corrected)", 
       x="CERES Score (Batch Corrected)", 
       y="Frequency") +
  theme(legend.position="none")

plt = grid.arrange(p1, p2)
savepath = "D:/Analysis/EMT/figures/CRISPR/batch_distribution.png"
ggsave(file=savepath, plt)
```

# 4. Get gene list that are universally lethal across all cell lines
This gene list was determined from Project Score, which prioritized CRISPR Scores using the Adaptive Daisy model. The publication to reference is Behan, F. M., Iorio, F., Picco, G., Gonçalves, E., Beaver, C. M., Migliardi, G., ... & Garnett, M. J. (2019). Prioritization of cancer therapeutic targets using CRISPR–Cas9 screens. Nature, 568(7753), 511-516.
```{r}
project_score_file = "D:/Data/CRISPR/Score/41586_2019_1103_MOESM4_ESM/Supplementary Table 3 - Core-Fitness and Context-Specificity Summary_v2.xlsx"
sheet_name = "Filter Genes"
project_score = read_excel(project_score_file, sheet_name)
```

In this method, core fitness genes were scored as 0, and they were designated as high chance for false positive genes. While we can perform a more empircal analysis to determine the threshold used to filter promiscuous genes versus specific essential genes, I have used the arbitrary threshold of a score of 50 to separate out core essential versus specific genes.
```{r}
specific_scores = project_score[project_score$`Dependency Score` > 50, ]
essential_scores = project_score[project_score$`Dependency Score` < 50, ]
```

Finally, I would like to see how the scores are distributed.
```{r, fig.height=7.5, fig.width=10, dpi=100}
tmp = reshape2::melt(project_score$`Dependency Score`)
p2 = ggplot(tmp, aes(x=value)) +  
  geom_histogram(bins=100) +
  labs(title="Project Score distribution", 
       x="Specificity Scores", 
       y="Frequency") +
  theme(legend.position="none")


savepath = "D:/Analysis/EMT/figures/CRISPR/project_score.png"
ggsave(file=savepath, p2)
```

Interestingly, most Scores are ubiquitous across all cancers. These maps will be used in the heatmaps below.

# 5. Grab NSCLC Lung Cancer Cell Line Data
The `cellmatch` dataframe contains the metadata for the `match_data`. We're interested in specifically analyzing NSCLC data.

## A. Subset the Data
```{r}
NSCLC_meta = cellmatch[cellmatch["lineage_subtype"]=="NSCLC", ]
NSCLC_meta = NSCLC_meta[!is.na(NSCLC_meta$primary_or_metastasis), ]
``` 

## B. Visualize Primary V. Metastatic
Let's visualize how many cells are primary site versus metastatic cancers. First, let's get the counts.
```{r}
unique(NSCLC_meta$primary_or_metastasis)
tmp = NSCLC_meta %>%
  group_by(primary_or_metastasis) %>%
  count()

tmp2 = cellmatch %>%
  group_by(primary_or_metastasis) %>%
  count()
tmp2= tmp2[-nrow(tmp2), ]
```

Now let's make the barplot.
```{r, fig.width=5, fig.height=5, dpi=100}
ggplot(data=tmp, aes(x=primary_or_metastasis, 
                          y=n, 
                          fill=primary_or_metastasis)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("Metastasis"="#EF476F", "Primary"="#073B4C")) +
  labs(title="NSCLC cell line distribution", 
       x="Type of NSCLC cell line", 
       y="Frequency") +
  theme(legend.position="none")

ggsave(filename="D:/Analysis/EMT/figures/Supporting Figures/NSCLC_Primary_v_Metastatic.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

ggplot(data=tmp2, aes(x=primary_or_metastasis, 
                          y=n, 
                          fill=primary_or_metastasis)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("Metastasis"="#EF476F", "Primary"="#073B4C")) +
  labs(title="All Primary/Metastatic Cancer Cell Line Distribution", 
       x="Type of NSCLC cell line", 
       y="Frequency") +
  theme(legend.position="none")

ggsave(filename="D:/Analysis/EMT/figures/Supporting Figures/All_Primary_v_Metastatic.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)
```
While the data is not skewed that much, it is still notable that there are more metastatic versus primary site NSCLC cell lines in the data.

## C. Grab NSCLC specific data
Finally, we'll grab the CERES scores and the FDR associated with each NSCLC cell line.
```{r}
NSCLC_ceres = match_data[, colnames(match_data) %in% NSCLC_meta$DepMap_ID]
NSCLC_fdr = padjmat[,  colnames(padjmat) %in% NSCLC_meta$DepMap_ID]
```

## D. Grab RECON1 metabolic reconstruction genes in the NSCLC dataset
To benchmark RECON1 results, we'll grab the CERES score gene entries that intersect with the metabolic reconstruction.
```{r}
# Grab map between Entrez and Symbol in RECON1
ids = rownames(NSCLC_ceres)
idx = ids %in% recon1_genes$Symbol

recon1_nsclc_ceres = NSCLC_ceres[idx, ]
recon1_nsclc_fdr = NSCLC_fdr[idx, ]
```

# 6. Visualize NSCLC Reactions of Interest from COBRA
Note that to translate reaction IDs --> gene IDs, I used the `findGenesFromRxns` function in the COBRA Toolbox.

## A. Bulk Reaction KO Sensitive Reactions
This module visualizes the reactions listed in the `make_manuscript_figures.Rmd` file for the bulk reaction KOs.

### i. Load Data
```{r}
# Top 50 reaction knockouts
top50_path = "D:/Analysis/EMT/data/top50_ko_map.csv"
top50_map = read.csv(top50_path,
                          header=FALSE)

# High confidence knockouts
highconf_path = "D:/Analysis/EMT/data/highconf_ko_map.csv"
highconf_map = read.csv(highconf_path,
                          header=FALSE)

# Low confidence knockouts
lowconf_path = "D:/Analysis/EMT/data/lowconf_ko_map.csv"
lowconf_map = read.csv(lowconf_path,
                          header=FALSE)
```

### ii. Get intersecting CRISPR data
```{r}
top50_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% top50_map$V1, ]
highconf_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% highconf_map$V1, ]
lowconf_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% lowconf_map$V1, ]
```

### iii. Visualize boxplots
In this case, I am visualizing boxplots of primary versus metastatic gene distributions that are important to this story. 

First, let's create the function that will make these boxplots.
```{r}
essentiality_boxplots = function(df, metadata){
  # Map melted dataframe to metadata
  tmp = reshape2::melt(df)
  tmp = merge(tmp, metadata, 
              by.x='Var2', by.y='DepMap_ID')
  
  # Create Boxplots
  plt = tmp %>%
          drop_na() %>%
          ggplot(aes(x=Var1, y=value, color=primary_or_metastasis)) +
          geom_boxplot(outlier.shape=NA) +
          geom_point(position=position_jitterdodge(), alpha=0.50) + 
          labs( x="Metabolic genes predicted by COBRA", 
               y="CERES Score",
               legend="Status") +
          theme(axis.text.x = element_text(angle=90, 
                                           hjust=1,
                                           vjust=0.3)) +
          geom_hline(yintercept=0,
                     color = "black",
                     size=0.5) +
          scale_color_manual(name="Cancer Status", 
                             values=c("#EF476F", "#073B4C")) +
          scale_alpha(guide="none")
  
  return(plt)
  
}
```


Note: the horizontal line at a CERES Score = 0 denotes no essentiality. 
```{r, fig.height=5, fig.width=10, dpi=100}

plt = essentiality_boxplots(top50_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/NSCLC/top50_ko_nsclc.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Low Confidence Reactions
plt = essentiality_boxplots(lowconf_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/NSCLC/lowconf_ko_nsclc.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)
```

## B. Bulk Overactive Metabolic Reactions
This module visualizes the reactions listed in the `make_manuscript_figures.Rmd` file for the overactive metabolic reactions.

### i. Load Data
```{r}
# Top 50 reaction knockouts
top50_path = "D:/Analysis/EMT/data/top50_flux_map.csv"
top50_map = read.csv(top50_path,
                     header=FALSE)

# Glycolysis
glycolysis_path = "D:/Analysis/EMT/data/glycolysis_flux_map.csv"
glycolysis_map = read.csv(glycolysis_path,
                          header=FALSE)

# Pentose Phosphate Pathway
ppp_path = "D:/Analysis/EMT/data/ppp_flux_map.csv"
ppp_map = read.csv(ppp_path,
                          header=FALSE)

# Specific Reactions
specific_path = "D:/Analysis/EMT/data/specific_flux_map.csv"
specific_map = read.csv(specific_path,
                          header=FALSE)
```

### ii. Get intersecting CRISPR data
```{r}
top50_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% top50_map$V1, ]
glycolysis_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% glycolysis_map$V1, ]
ppp_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% ppp_map$V1, ]
specific_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% specific_map$V1, ]
```

### iii. Visualize heatmap
```{r, fig.height=5, fig.width=10, dpi=100}
# Top 50 Metabolically Active Reactions
plt = essentiality_boxplots(top50_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/NSCLC/top50_flux_nsclc.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Glycolytic Reactions
plt = essentiality_boxplots(glycolysis_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/NSCLC/glycolysis_flux_nsclc.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Pentose Phosphate Pathway Reactions
plt = essentiality_boxplots(ppp_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/NSCLC/ppp_flux_nsclc.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Specific Reactions
plt = essentiality_boxplots(specific_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/NSCLC/specific_flux_nsclc.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)
```

## C. Single-Cell KO Sensitive Reactions
This module visualizes the reactions listed in the `make_manuscript_figures.Rmd` file for the single-cell reaction knockouts.

### i. Load Data
```{r}
# Top 50 reaction knockouts
singlecell_path = "D:/Analysis/EMT/data/singlecell_ko_map.csv"
singlecell_map = read.csv(singlecell_path,
                          header=FALSE)
```

### ii. Get intersecting CRISPR data
```{r}
singlecell_ceres = recon1_nsclc_ceres[row.names(recon1_nsclc_ceres) %in% singlecell_map$V1, ]
```

### iii. Visualize heatmap
```{r, fig.height=5, fig.width=10, dpi=100}
plt = essentiality_boxplots(singlecell_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/NSCLC/singlecell_ko_nsclc.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)
```

# 7. Visualize All CERES Score Mapped to Reactions of Interest
## A. Bulk Reaction KO Sensitive Reactions
This module visualizes the reactions listed in the `make_manuscript_figures.Rmd` file for the bulk reaction KOs.

### i. Load Data
```{r}
# Top 50 reaction knockouts
top50_path = "D:/Analysis/EMT/data/top50_ko_map.csv"
top50_map = read.csv(top50_path,
                          header=FALSE)

# High confidence knockouts
highconf_path = "D:/Analysis/EMT/data/highconf_ko_map.csv"
highconf_map = read.csv(highconf_path,
                          header=FALSE)

# Low confidence knockouts
lowconf_path = "D:/Analysis/EMT/data/lowconf_ko_map.csv"
lowconf_map = read.csv(lowconf_path,
                          header=FALSE)
```

### ii. Get intersecting CRISPR data
```{r}
top50_ceres = match_data[row.names(match_data) %in% top50_map$V1, ]
highconf_ceres = match_data[row.names(match_data) %in% highconf_map$V1, ]
lowconf_ceres = match_data[row.names(match_data) %in% lowconf_map$V1, ]
```

### iii. Visualize boxplots
In this case, I am visualizing boxplots of primary versus metastatic gene distributions that are important to this story. 

Note: the horizontal line at a CERES Score = 0 denotes no essentiality. 
```{r, fig.height=5, fig.width=10, dpi=100}
# Top 50 Reactions
plt = essentiality_boxplots(top50_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/All/top50_ko_all.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Low Confidence Reactions
plt = essentiality_boxplots(lowconf_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/All/lowconf_ko_all.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)
```

## B. Bulk Overactive Metabolic Reactions
This module visualizes the reactions listed in the `make_manuscript_figures.Rmd` file for the overactive metabolic reactions.

### i. Load Data
```{r}
# Top 50 reaction knockouts
top50_path = "D:/Analysis/EMT/data/top50_flux_map.csv"
top50_map = read.csv(top50_path,
                     header=FALSE)

# Glycolysis
glycolysis_path = "D:/Analysis/EMT/data/glycolysis_flux_map.csv"
glycolysis_map = read.csv(glycolysis_path,
                          header=FALSE)

# Pentose Phosphate Pathway
ppp_path = "D:/Analysis/EMT/data/ppp_flux_map.csv"
ppp_map = read.csv(ppp_path,
                          header=FALSE)

# Specific Reactions
specific_path = "D:/Analysis/EMT/data/specific_flux_map.csv"
specific_map = read.csv(specific_path,
                          header=FALSE)
```

### ii. Get intersecting CRISPR data
```{r}
top50_ceres = match_data[row.names(match_data) %in% top50_map$V1, ]
glycolysis_ceres = match_data[row.names(match_data) %in% glycolysis_map$V1, ]
ppp_ceres = match_data[row.names(match_data) %in% ppp_map$V1, ]
specific_ceres = match_data[row.names(match_data) %in% specific_map$V1, ]
```

### iii. Visualize heatmap
```{r, fig.height=5, fig.width=10, dpi=100}
# Top 50 Metabolically Active Reactions
plt = essentiality_boxplots(top50_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/All/top50_flux_all.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Glycolytic Reactions
plt = essentiality_boxplots(glycolysis_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/All/glycolysis_flux_all.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Pentose Phosphate Pathway Reactions
plt = essentiality_boxplots(ppp_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/All/ppp_flux_all.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)

# Specific Reactions
plt = essentiality_boxplots(specific_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/All/specific_flux_all.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)
```

## C. Single-Cell KO Sensitive Reactions
This module visualizes the reactions listed in the `make_manuscript_figures.Rmd` file for the single-cell reaction knockouts.

### i. Load Data
```{r}
# Top 50 reaction knockouts
singlecell_path = "D:/Analysis/EMT/data/singlecell_ko_map.csv"
singlecell_map = read.csv(singlecell_path,
                          header=FALSE)
```

### ii. Get intersecting CRISPR data
```{r}
singlecell_ceres = match_data[row.names(match_data) %in% singlecell_map$V1, ]
```

### iii. Visualize heatmap
```{r, fig.height=5, fig.width=10, dpi=100}
plt = essentiality_boxplots(singlecell_ceres, NSCLC_meta)
ggsave(filename="D:/Analysis/EMT/figures/CRISPR/All/singlecell_ko_all.png",
       plot=last_plot(),
       device=png(),
       width=10, 
       height=5,
       dpi=100)
```

# 8. Visualize CERES Score Data
Now that we have the entire subset, let's visualize the data by splitting the data between primary and metastatic cell lines. 

First, we'll separate the data into metastatic and primary.
```{r}
met_idx = NSCLC_meta$primary_or_metastasis == "Metastasis"
prim_idx = NSCLC_meta$primary_or_metastasis == "Primary"
```

Then we'll pre-process some matrices for visualization via heatmap.
```{r}
# Metastatic data
met_data = recon1_nsclc_ceres[, met_idx]
met_fdr = recon1_nsclc_fdr[, met_idx]
met_bin = met_fdr > 0.05
colnames(met_data) = NSCLC_meta$stripped_cell_line_name[met_idx]

# Turn non-significant datapoints into black
masked_met_data = met_data
masked_met_data[met_bin] = NA

# Primary site data
prim_data = recon1_nsclc_ceres[, prim_idx]
prim_fdr = recon1_nsclc_fdr[, prim_idx]
prim_bin = prim_fdr > 0.05
colnames(prim_data) = NSCLC_meta$stripped_cell_line_name[prim_idx]

# Turn non-significant datapoints into black
masked_prim_data = prim_data
masked_prim_data[prim_bin] = NA
```

Finally, filter by essential / specifically lethal.
```{r}
essential_met = met_data[row.names(met_data) %in% essential_scores$`HGNC Symbol`, ]
specific_met = met_data[row.names(met_data) %in% specific_scores$`HGNC Symbol`, ]
essential_masked_met = masked_met_data[row.names(masked_met_data) %in% essential_scores$`HGNC Symbol`, ]
specific_masked_met = masked_met_data[row.names(masked_met_data) %in% specific_scores$`HGNC Symbol`, ]

essential_prim = prim_data[row.names(prim_data) %in% essential_scores$`HGNC Symbol`, ]
specific_prim = prim_data[row.names(prim_data) %in% specific_scores$`HGNC Symbol`, ]
essential_masked_prim = masked_prim_data[row.names(masked_prim_data) %in% essential_scores$`HGNC Symbol`, ]
specific_masked_prim = masked_prim_data[row.names(masked_prim_data) %in% specific_scores$`HGNC Symbol`, ]
```

### i. Create annotation bar for primary/metastatic
Let's create the column annotation bars.
```{r}
h1_col = HeatmapAnnotation(hm1=anno_block(gp=gpar(fill="#EF476F"),
                                      labels=c("Metastatic"), 
                                      labels_gp=gpar(col="black", 
                                                     fontsize=10)))

h2_col = HeatmapAnnotation(hm2=anno_block(gp=gpar(fill="#073B4C"),
                                      labels=c("Primary"), 
                                      labels_gp=gpar(col="white", 
                                                     fontsize=10)))
```

### ii. Create annotation bar for essential/specific
```{r}
h1_row = rowAnnotation(hm1=anno_block(gp=gpar(fill="#EF476F"),
                                      labels=c("Essential"), 
                                      labels_gp=gpar(col="black", 
                                                     fontsize=10)))

h2_row = rowAnnotation(hm2=anno_block(gp=gpar(fill="#073B4C"),
                                      labels=c("Specific"), 
                                      labels_gp=gpar(col="white", 
                                                     fontsize=10)))
```

### ii. Visualize all data
This heatmap contains all 1400+ metabolic genes across 44 cell lines.
```{r, fig.width=12.5, fig.height=40, fig.align='right', dpi=100}
# Metastatic heatmap
hm1 = Heatmap(as.matrix(essential_met),
              left_annotation=h1_row,
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=TRUE,
              show_row_names = FALSE,
              show_row_dend=FALSE,
              top_annotation=h1_col, 
              column_names_side = "top",
              width = ncol(essential_met)*unit(3, "mm"), 
              height = nrow(essential_met)*unit(1, "mm"),
              show_heatmap_legend=FALSE)
idx1 = row_order(hm1)

hm2 = Heatmap(as.matrix(specific_met),
              left_annotation=h2_row,
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=TRUE,
              show_row_dend=FALSE,
              show_row_names = FALSE,
              width = ncol(specific_met)*unit(3, "mm"), 
              height = nrow(specific_met)*unit(1, "mm"),
              show_heatmap_legend=FALSE,
              show_column_names = FALSE)
idx2 = row_order(hm2)


# Primary heatmap
hm3 = Heatmap(as.matrix(essential_prim)[idx1, ],
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              top_annotation=h2_col, 
              column_names_side = "top",
              width = ncol(essential_prim)*unit(3, "mm"), 
              height = nrow(essential_prim)*unit(1, "mm"),
              heatmap_legend_param=list(title="CERES Score"))

hm4 = Heatmap(as.matrix(specific_prim)[idx2, ],
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              width = ncol(specific_prim)*unit(3, "mm"), 
              height = nrow(specific_prim)*unit(1, "mm"),
              show_heatmap_legend=FALSE,
              show_column_names = FALSE)

metastatic = hm1 %v% hm2
primary = hm3 %v% hm4
```

```{r}
library(cowplot)
plot_grid(hm1, hm2, hm3, hm4, ncol=2)
lyot = grid.layout(nrow=1, ncol=2, widths=unit(0.1, "null"))
grid.newpage()
pushViewport(viewport(layout = lyot))
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 1, just="bottom"))
draw(metastatic, newpage=FALSE)
upViewport()

pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 2, just="bottom"))
draw(primary, newpage=FALSE)
upViewport()
```

### iii. Visualize the signficant data only
We'll black out the non-significant entries based on FDR > 0.05.
```{r, fig.width=12.5, fig.height=40, fig.align='right', dpi=100}
# Metastatic heatmap
hm5 = Heatmap(as.matrix(essential_masked_met[idx1, ]),
              left_annotation=h1_row,
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=FALSE,
              show_row_names = FALSE,
              show_row_dend=FALSE,
              top_annotation=h1_col, 
              column_names_side = "top",
              width = ncol(essential_masked_met)*unit(3, "mm"), 
              height = nrow(essential_masked_met)*unit(1, "mm"),
              show_heatmap_legend=FALSE)

hm6 = Heatmap(as.matrix(specific_masked_met[idx2, ]),
              left_annotation=h2_row,
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=FALSE,
              show_row_dend=FALSE,
              show_row_names = FALSE,
              width = ncol(specific_masked_met)*unit(3, "mm"), 
              height = nrow(specific_masked_met)*unit(1, "mm"),
              show_heatmap_legend=FALSE,
              show_column_names = FALSE)


# Primary heatmap
hm7 = Heatmap(as.matrix(essential_masked_prim)[idx1, ],
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              top_annotation=h2_col, 
              column_names_side = "top",
              width = ncol(essential_masked_prim)*unit(3, "mm"), 
              height = nrow(essential_masked_prim)*unit(1, "mm"),
              heatmap_legend_param=list(title="CERES Score"))

hm8 = Heatmap(as.matrix(specific_masked_prim)[idx2, ],
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              width = ncol(specific_masked_prim)*unit(3, "mm"), 
              height = nrow(specific_masked_prim)*unit(1, "mm"),
              show_heatmap_legend=FALSE,
              show_column_names = FALSE)
```

```{r}
#hm5
#hm6
#hm7
#hm8
metastatic = hm5 %v% hm6
metastatic
primary = hm7 %v% hm8
primary
#plot_grid(hm5, hm6, hm7, hm8, ncol=2)
#lyot = grid.layout(nrow=1, ncol=2, widths=unit(0.1, "null"))
grid.newpage()
pushViewport(viewport(layout = lyot))
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 1, just="bottom"))
draw(metastatic, newpage=FALSE)
upViewport()

pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 2, just="bottom"))
draw(primary, newpage=FALSE)
upViewport()
```

### iv. Visualize the bottom 50 metabolic genes
Now let's zoom in on the bottom 50 metabolic genes.
```{r, fig.width=12.5, fig.height=15, fig.align='right', dpi=100}
tmp1 = as.matrix(essential_masked_met)[idx1, ]
tmp1 = tmp1[(nrow(tmp1)-50):nrow(tmp1), ]

tmp2 = as.matrix(specific_masked_met)[idx2, ]
tmp2 = tmp2[(nrow(tmp2)-50):nrow(tmp2), ]

tmp3 = as.matrix(essential_masked_prim)[idx1, ]
tmp3 = tmp3[(nrow(tmp3)-50):nrow(tmp3), ]

tmp4 = as.matrix(specific_masked_prim)[idx2, ]
tmp4 = tmp4[(nrow(tmp4)-50):nrow(tmp4), ]

# Metastatic heatmap
hm9 = Heatmap(as.matrix(tmp1),
              left_annotation=h1_row,
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=FALSE,
              show_row_names = FALSE,
              show_row_dend=FALSE,
              top_annotation=h1_col, 
              column_names_side = "top",
              width = ncol(tmp1)*unit(3, "mm"), 
              height = nrow(tmp1)*unit(3, "mm"),
              show_heatmap_legend=FALSE)

hm10 = Heatmap(as.matrix(tmp2),
              left_annotation=h2_row,
              na_col = "black",
              cluster_columns=FALSE,
              cluster_rows=FALSE,
              show_row_dend=FALSE,
              show_row_names = FALSE,
              width = ncol(tmp2)*unit(3, "mm"), 
              height = nrow(tmp2)*unit(3, "mm"),
              show_heatmap_legend=FALSE,
              show_column_names = FALSE)


# Primary heatmap
hm11 = Heatmap(as.matrix(tmp3),
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              top_annotation=h2_col, 
              column_names_side = "top",
              width = ncol(tmp3)*unit(3, "mm"), 
              height = nrow(tmp3)*unit(3, "mm"),
              heatmap_legend_param=list(title="CERES Score"))

hm12 = Heatmap(as.matrix(tmp4),
              na_col = "black",
              cluster_columns=FALSE, 
              cluster_rows=FALSE,
              width = ncol(tmp4)*unit(3, "mm"), 
              height = nrow(tmp4)*unit(3, "mm"),
              show_heatmap_legend=FALSE,
              show_column_names = FALSE)
```

```{r, fig.width=12.5, fig.height=15, fig.align='right', dpi=300}
#hm5
#hm6
#hm7
#hm8
metastatic = hm9 %v% hm10
metastatic
primary = hm11 %v% hm12
primary
#plot_grid(hm5, hm6, hm7, hm8, ncol=2)
lyot = grid.layout(nrow=1, ncol=2)
grid.newpage()
pushViewport(viewport(layout = lyot))
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 1, just="bottom"))
draw(metastatic, newpage=FALSE)
upViewport()

pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 2, just="bottom"))
draw(primary, newpage=FALSE)
upViewport()
```

Now that we have a sense of what metabolic genes are generally essential between the metastatic and primary cell lines, we'll see which ones intersect between the model predictions and the bottom 50 data.

# 5. Get intersecting data with RECON1 simulations

# 7. Get differentially expressed CERES score by comparing metastatic / primary site using limma-voom

## Data preprocessing
Format the dataset
```{r}
sample_info = merged_lung[, 1:5]
data = merged_lung[, -c(1:5)]
rownames(data) = sample_info[, 2]
data = t(data)
sample_info = sample_info %>% dplyr::select(2:3)
```

Now we'll floor positive CERES Scores to 0, and take the absolute value for the CERES Scores
```{r}
data[data > 0] = 0
data = abs(data)
data[is.na(data)] = 0
```

## Preparing limma-voom data 
Now let's preprocess the data
```{r}
# Create a DGE List data object
data = cpm(data)
d0 = DGEList(data)
d0 = calcNormFactors(d0)

# Filter out bad data points
cutoff = 1
drop = which(apply(cpm(d0), 1, max) < cutoff)
d = d0[-drop,] 
dim(d)
```

We need to set up the groups we want to compare and the design matrix.
```{r}
group = as.factor(sample_info$primary_or_metastasis)
design = model.matrix(~0+group)
```

## Run limma-voom
Now let's run *limma-voom*
```{r}
v = voom(d, design, plot=TRUE)
fit = lmFit(v, design, method='robust')
result = eBayes(fit, robust=TRUE)
```

Finally, let's get the top 20 results
```{r}
top.table = topTable(result)
head(top.table, 20)
write.table(top.table, "clipboard", sep="\t", col.names=TRUE)
```



# 9. Get differentially expressed genes for A549 from CERES score
First, we'll get the row corresponding to the cancer cell line A549.
```{r}
a549_ceres = merged_lung[merged_lung$stripped_cell_line_name=="A549",]
a549_ceres[2, ] = merged_pval[merged_pval$stripped_cell_line_name=="A549",]
a549_ceres = a549_ceres[, 6:ncol(a549_ceres)]
a549_ceres = t(a549_ceres)
a549_ceres = data.frame(a549_ceres)
colnames(a549_ceres) = c("CERES", "PValue")
```

Now, let's grab the metabolic genes that are in RECON1
```{r}
xname = data.frame(entrez=rownames(a549_ceres))
tmp = merge(xname, recon1_match, by='entrez')
tmp[, 3:4] = NULL
tmp = unique(tmp)
a549_ceres$Symbol = tmp$symbol
```

## Plot Venn Diagram
```{r}
gene_overlap = list(
  A549_specific = sig_a549_ceres$Symbol,
  Core_fitness = genes_to_filter$`HGNC Symbol`
)
ggVennDiagram(gene_overlap)
```

Finally, we'll grab the genes that have a left-tailed p-value less than 0.05
```{r}
sig_a549_ceres = a549_ceres[a549_ceres$PValue < 0.05, ]
genes_to_filter = genes_to_filter$`Entrez ID`
genes_to_filter = data.frame(genes_to_filter)
ubiq_genes = merge(genes_to_filter, 
                   sig_a549_ceres, 
                   by.x='genes_to_filter', 
                   by.y="row.names")
unique_genes = sig_a549_ceres[!sig_a549_ceres$Symbol %in% ubiq_genes$Symbol, ]
write.table(sig_a549_ceres, "clipboard", sep="\t", col.names=TRUE)
```

# 10. Identify where the predicted knockouts are in the CERES distribution
First, let's load the model predictions
```{r}
recon1_prediction = "D:/Analysis/EMT/EMT_RECON1_KO_PRED.xlsx"
pred = read_excel(recon1_prediction)
a549_ceres$is_pred = rownames(a549_ceres) == pred$Entrez
a549_ceres$category = "A549"
sum(a549_ceres$is_pred)
```

Now let's create the histogram
```{r}
# Create histogram of reaction distribution
p1 = ggplot(a549_ceres, aes(x=CERES, fill=is_pred)) +
  geom_histogram(bins=100, show.legend=TRUE) +
  theme_minimal() +
  theme(legend.position="none") +
  xlim(-0.5, 0.5) +
  theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_blank(),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin()) +
  labs(title="",
       x=element_blank(),
       y="Frequency") + 
     scale_color_manual(values=c("TRUE" = "red", 
                                 "FALSE" = "gray"),
                        aesthetics = c("colour", "fill"))
```

Let's create the scatter plot
```{r}
p2 = ggplot(a549_ceres, aes(x=CERES, y=category, color=is_pred)) +
     geom_point(show.legend=FALSE, alpha=0.01) +
     geom_jitter() +
     theme_minimal() + 
     theme(legend.position="none") +
     xlim(-0.5, 0.5) +
     labs(x="CERES Score",
          y=element_blank()) + 
     scale_color_manual(values=c("TRUE" = "red", 
                                 "FALSE" = "gray"),
                        aesthetics = c("colour", "fill"))
```

Finally construct the figure.
```{r}
figure = ggarrange(p1, p2,
                   ncol = 1, nrow = 2)
figure
```