---
title: "Single Cell Differential Gene Expression Analysis"
author: Scott Campit
output: html_notebook
---

# Differential expression
This code block gets differentially expressed genes from the preprocessed Seurat object.

## Bulk scRNASeq across time points

Cook et al performed a GAM with loess on the time covariate for differential expression. Below is the modified `diffExp` function that takes in the Seurat object to get the result.

```{r}
diffExp = function(seurat_object){
  cells_keep = colnames(seurat_object)[-which(seurat_object$Time %in% 
                                                c("8h_rm", "1d_rm", "3d_rm"))]
  subset_seurat = subset(seurat_object, cells=cells_keep)
  subset_seurat$Timepoint = as.numeric(gsub("[^0-9.-]", 
                                            "", 
                                            subset_seurat$Time))
  subset_seurat$Timepoint[subset_seurat$Timepoint==8] = 0.33 
  
  exp = as.matrix(GetAssayData(subset_seurat, slot="data")) 
  
  # Fit natural splines to the dataset
  gam = apply(exp, 1, function(x){
    data = data.frame(exp=x, timepoint=subset_seurat$Timepoint)
    temp = gam(exp ~ ns(timepoint, df=3), data=data)
    res = data.frame(Intercept=temp$coefficients[1],
                      Spline_1=temp$coefficients[2],
                      Spline_2=temp$coefficients[3],
                      Spline_3=temp$coefficients[4],
                      pval=summary(temp)$p.table[,4])
  })
  
  # Compute adjusted P-value using FDR
  results <- do.call("rbind", gam)
  results$qval <- p.adjust(results$pval, method="fdr")
  results$Gene <- rownames(results)
  return(results)
}
```

Compute DGE for A549 TGF treatment and save it to a csv file.
```{r}
# Perform DEA
a549_tgfb1_dge = diffExp(a549)

# Write to file
filepath = "~/Data/Expression/GSE147405/dge_a549_tgfb1.csv"
write.csv(a549_tgfb1_dge, file=filepath, quote=F)

```

Let's now get a list of significant genes

```{r}
a549_tgfb1_sig <- filter(a549_tgfb1_dge, qval <= 0.05)

# Write to file
filepath = "~/Data/Expression/GSE147405/a549_tgfb1_sig.csv"
write.csv(a549_tgfb1_sig, file=filepath, quote=F)

```


```{r}
library(preprocessCore)
exp = as.matrix(GetAssayData(a549, slot="scale.data"))
exp = exp - median(exp)
exp$Symbol = rownames(exp)
```
## Pseudotime analysis

```{r}
plot_data <- a549@meta.data
plot_data$UMAP1 <- Embeddings(a549, "umap_pseudo")[,1]
plot_data$UMAP2 <- Embeddinginstall.packages(c("reshape2", "ggplot2"))
plot_data$UMAP2 <- Embeddings(a549, "umap_pseudo")[,2]

```

```{r}
PRGn <- rev(RColorBrewer::brewer.pal(11, "PRGn"))
color_ramp <- c(PRGn[1], PRGn[2], PRGn[3], PRGn[4], "grey50", PRGn[7], PRGn[8], PRGn[9])
```


```{r}
# Actual time points
timepoint_plot <- ggplot(plot_data, aes(x=UMAP1, y=UMAP2)) +
  geom_point(size=2, alpha=0.75, aes(color=Sample),
             shape=16) +
  scale_color_manual(values=color_ramp) +
  theme_void() +
  theme(legend.position="none")

# Pseudotime
pseudo_plot <- ggplot(plot_data, aes(x=UMAP1, y=UMAP2)) +
  geom_point(size=2, alpha=0.75, aes(color=Pseudotime),
             shape=16) +
  scale_color_viridis(option="D") +
  theme_void() +
  theme(legend.title=element_blank(),
        legend.text=element_blank())
```

```{r}
timepoint_plot
pseudo_plot
```

```{r}
mlt <- melt(exp)
ggplot(data = mlt, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile()
```